<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>第9章 访问NCBI Entrez数据库 &mdash; biopython_cn 0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="biopython_cn 0.1 documentation" href="index.html" />
    <link rel="next" title="第10章 Swiss-Prot和ExPASy" href="chr10.html" />
    <link rel="prev" title="第8章 BLAST和其他序列搜索工具(实验性质的代码)" href="chr08.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="chr10.html" title="第10章 Swiss-Prot和ExPASy"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="chr08.html" title="第8章 BLAST和其他序列搜索工具(实验性质的代码)"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">biopython_cn 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="ncbi-entrez">
<h1>第9章  访问NCBI Entrez数据库<a class="headerlink" href="#ncbi-entrez" title="Permalink to this headline">¶</a></h1>
<p>Entrez
(<a class="reference external" href="http://www.ncbi.nlm.nih.gov/Entrez">http://www.ncbi.nlm.nih.gov/Entrez</a>)
是一个给客户提供NCBI各个数据库（如PubMed, GeneBank, GEO等等）访问的检索系统。
用户可以通过浏览器手动输入查询条目访问Entrez，也可以使用Biopython的 <tt class="docutils literal"><span class="pre">Bio.Entrez</span></tt> 模块以编程方式访问来访问Entrez。
如果使用第二种方法，用户用一个Python脚本就可以实现在PubMed里面搜索或者从GenBank下载数据。</p>
<p><tt class="docutils literal"><span class="pre">Bio.Entrez</span></tt> 模块利用了Entrez Programming Utilities（也称作EUtils），包含八个工具，详情请见NCBI的网站：
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/utils/">http://www.ncbi.nlm.nih.gov/entrez/utils/</a>.
每个工具都能在Python的 <tt class="docutils literal"><span class="pre">Bio.Entrez</span></tt> 模块中找到对应函数，后面会详细讲到。这个模块可以保证用来查询的URL
的正确性，并且向NCBI要求的一样，每三秒钟查询的次数不超过一。</p>
<p>EUtils返回的输出结果通常是XML格式，我们有以下不同的方法来解析这种类型的输入文件：</p>
<ol class="arabic simple">
<li>使用 <tt class="docutils literal"><span class="pre">Bio.Entrez</span></tt>解析器将XML输出的解析成Python对象;</li>
<li>使用Python标准库中的DOM (Document Object Model)解析器;</li>
<li>使用Python标准库中的SAX (Simple API for XML)解析器;</li>
<li>把XML输出当做原始的文本文件，通过字符串查找和处理来进行解析；</li>
</ol>
<p>对于DOM和SAX解析器，可以查看Python的文档. <tt class="docutils literal"><span class="pre">Bio.Entrez</span></tt> 中使用到的解析器将会在下面讨论.</p>
<p>NCBI使用DTD (Document Type Definition)文件来描述XML文件中所包含信息的结构. 大多数NCBI使用的DTD文件
格式都包含在了Biopython发行包里。当NCBI Entrez读入一个XML格式的文件的时候，<tt class="docutils literal"><span class="pre">Bio.Entrez</span></tt>
将会使用DTD文件。</p>
<p>有时候，你可能会发现与某种特殊的XML相关的DTD文件在Biopython发行包里面不存在。当NCBI升级它的
DTD文件的时候，这种情况可能发生。如果发生这种情况，<tt class="docutils literal"><span class="pre">Entrez.read</span></tt> 将会显示丢失的DTD文件名字和URL的
警示信息。解析器会通过互联网获取缺失的DTD文件，让XML的分析继续正常进行。如果本地存在对应的DTD文件的
话，处理起来会更快。因此，为了更快的处理，我们可以通过警示信息里面的URL来下载对应的DTD文件，将文件放在DTD
文件默认存放的文件夹 <tt class="docutils literal"><span class="pre">...site-packages/Bio/Entrez/DTDs</span></tt> 。如果你没有权限进入这个文件夹，你也可以把
DTD文件放到 <tt class="docutils literal"><span class="pre">~/.biopython/Bio/Entrez/DTDs</span></tt> 这个目录，<tt class="docutils literal"><span class="pre">~</span></tt> 表示的是你的Home目录。因为这个目录会先于
<tt class="docutils literal"><span class="pre">...site-packages/Bio/Entrez/DTDs</span></tt> 被解析器读取，所以当 <tt class="docutils literal"><span class="pre">...site-packages/Bio/Entrez/DTDs</span></tt>
下面的DTD文件过时的时候，你也可以将最新版本的DTD文件放到Home目录的那个文件夹下面。当然也有其他方案，如果你
是通过源码来安装的Biopython，你可以将DTD文件放到源码的 <tt class="docutils literal"><span class="pre">Bio/Entrez/DTDs</span></tt> 文件夹下，然后重新安装Biopython。
这样会将新的DTD文件和之前的一样地安装到正确的位置。</p>
<p>Entrez Programming Utilities也可以生成其他格式的输出文件，比如Fasta、序列数据库里面的GenBank文件格式
或者文献数据库里面的MedLine格式，更多内容将会在 <a class="reference external" href="#sec:entrez-specialized-parsers">9.12</a> 中讨论。</p>
<div class="section" id="entrez">
<h2>9.1  Entrez 简介<a class="headerlink" href="#entrez" title="Permalink to this headline">¶</a></h2>
<p>在我们通过Biopython访问NCBI的线上资源（通过 <tt class="docutils literal"><span class="pre">Bio.Entrez</span></tt> 或者其他模块）的时候，请先阅读 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/books/NBK25497/#chapter2.Usage_Guidelines_and_Requiremen">NCBI的Entrez
用户规范</a>.
如果NCBI发现你在滥用他们的系统，他们会禁止你的访问。</p>
<p>详细规范如下：</p>
<ul>
<li><p class="first">对任何连续超过100次的访问请求，请在周末时间或者避开美国的使用高峰时间。这个取决于你是否遵从。</p>
</li>
<li><p class="first">使用这个网址 <a class="reference external" href="http://eutils.ncbi.nlm.nih.gov">http://eutils.ncbi.nlm.nih.gov</a> ，
而不是通常的NCBI网址。Biopython使用的是这个网址。</p>
</li>
<li><p class="first">每秒钟不要超过三次请求（比2009年年初的每三秒钟最多一次请求要宽松）。这个由Biopython自动强制实行。</p>
</li>
<li><p class="first">使用email参数，这样如果遇到什么问题，NCBI可以通过邮件联系到你。你可以在每次请求Entrez的时候明确的设置
这个参数（例如，在参数列表中包含 <tt class="docutils literal"><span class="pre">email=&quot;A.N.Other&#64;example.com&quot;</span></tt> ），或者你也可以设置一个全局的email
地址：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>
</pre></div>
</div>
<p><tt class="docutils literal"><span class="pre">Bio.Entrez</span></tt> 将会在每次向Entrez请求的时候使用这个邮件地址。请千万不要胡乱的填写邮件地址，不填写都比
这要好。邮件的参数从2010年6月1日将是强制的参数。在过度使用的情况下，NCBI会在封锁用户访问E-utilities之前尝试通过
用户提供的邮件地址联系。</p>
</li>
<li><p class="first">如果你是在一个大的软件包里面使用Biopython的，请通过tool这个参数明确说明。你既可以在每次请求访问Entrez
的时候通过参数明确地指明使用的工具（例如，在参数列表中包含 <tt class="docutils literal"><span class="pre">tool=&quot;MyLocalScript&quot;</span></tt> ），或者你也可以
设置一个全局的tool名称：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">tool</span> <span class="o">=</span> <span class="s">&quot;MyLocalScript&quot;</span>
</pre></div>
</div>
<p>默认的tool名称是Biopython。</p>
</li>
<li><p class="first">对于大规模的查询请求，NCBI也推荐使用他们的会话历史特性（ WebEnv会话cookie字符串，见
章节 <a class="reference external" href="#sec:entrez-webenv">9.15</a> ）。 只是这个稍微有点复杂。</p>
</li>
</ul>
<p>最后，根据你的使用情况选择不同的策略。如果你打算下载大量的数据，最好使用其他的方法。比如，你想得到所有人的
基因的数据，那么考虑通过FTP得到每个染色体的GenBank文件，然后将这些文件导入到你自己的BioSQL数据库里面去。
(请见Section <a class="reference external" href="#sec:BioSQL">18.5</a>).</p>
</div>
<div class="section" id="einfo-entrez">
<h2>9.2  EInfo: 获取Entrez数据库的信息<a class="headerlink" href="#einfo-entrez" title="Permalink to this headline">¶</a></h2>
<p>EInfo为每个NCBI的数据库提供了条目索引，最近更新的时间以及可用的链接。此外，你可以很容易的使用EInfo通过
Entrez获取所有数据库名字的列表：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>     <span class="c"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">einfo</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">result</span> <span class="o">=</span> <span class="n">handle</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
</pre></div>
</div>
<p>变量 <tt class="docutils literal"><span class="pre">result</span></tt> 现在包含了XML格式的数据库列表：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">result</span>
<span class="go">&lt;?xml version=&quot;1.0&quot;?&gt;</span>
<span class="go">&lt;!DOCTYPE eInfoResult PUBLIC &quot;-//NLM//DTD eInfoResult, 11 May 2002//EN&quot;</span>
<span class="go"> &quot;http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd&quot;&gt;</span>
<span class="go">&lt;eInfoResult&gt;</span>
<span class="go">&lt;DbList&gt;</span>
<span class="go">        &lt;DbName&gt;pubmed&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;protein&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;nucleotide&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;nuccore&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;nucgss&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;nucest&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;structure&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;genome&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;books&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;cancerchromosomes&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;cdd&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;gap&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;domains&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;gene&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;genomeprj&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;gensat&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;geo&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;gds&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;homologene&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;journals&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;mesh&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;ncbisearch&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;nlmcatalog&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;omia&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;omim&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;pmc&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;popset&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;probe&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;proteinclusters&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;pcassay&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;pccompound&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;pcsubstance&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;snp&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;taxonomy&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;toolkit&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;unigene&lt;/DbName&gt;</span>
<span class="go">        &lt;DbName&gt;unists&lt;/DbName&gt;</span>
<span class="go">&lt;/DbList&gt;</span>
<span class="go">&lt;/eInfoResult&gt;</span>
</pre></div>
</div>
<p>因为这是一个相当简单的XML文件，我们可以简单的通过字符串查找提取里面所包含的信息。使用 <tt class="docutils literal"><span class="pre">Bio.Entrez</span></tt> 的解析器，
我们可以直接将这个XML读入到一个Python对象里面去：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">einfo</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
</pre></div>
</div>
<p>现在 <tt class="docutils literal"><span class="pre">record</span></tt> 是拥有一个确定键值的字典：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">[u&#39;DbList&#39;]</span>
</pre></div>
</div>
<p>这个键对应的值存储了上面XML文件里面包含的数据库名字的列表：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;DbList&quot;</span><span class="p">]</span>
<span class="go">[&#39;pubmed&#39;, &#39;protein&#39;, &#39;nucleotide&#39;, &#39;nuccore&#39;, &#39;nucgss&#39;, &#39;nucest&#39;,</span>
<span class="go"> &#39;structure&#39;, &#39;genome&#39;, &#39;books&#39;, &#39;cancerchromosomes&#39;, &#39;cdd&#39;, &#39;gap&#39;,</span>
<span class="go"> &#39;domains&#39;, &#39;gene&#39;, &#39;genomeprj&#39;, &#39;gensat&#39;, &#39;geo&#39;, &#39;gds&#39;, &#39;homologene&#39;,</span>
<span class="go"> &#39;journals&#39;, &#39;mesh&#39;, &#39;ncbisearch&#39;, &#39;nlmcatalog&#39;, &#39;omia&#39;, &#39;omim&#39;, &#39;pmc&#39;,</span>
<span class="go"> &#39;popset&#39;, &#39;probe&#39;, &#39;proteinclusters&#39;, &#39;pcassay&#39;, &#39;pccompound&#39;,</span>
<span class="go"> &#39;pcsubstance&#39;, &#39;snp&#39;, &#39;taxonomy&#39;, &#39;toolkit&#39;, &#39;unigene&#39;, &#39;unists&#39;]</span>
</pre></div>
</div>
<p>对于这些数据库，我们可以使用EInfo获得更多的信息：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">einfo</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;pubmed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;DbInfo&quot;</span><span class="p">][</span><span class="s">&quot;Description&quot;</span><span class="p">]</span>
<span class="go">&#39;PubMed bibliographic record&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;DbInfo&quot;</span><span class="p">][</span><span class="s">&quot;Count&quot;</span><span class="p">]</span>
<span class="go">&#39;17989604&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;DbInfo&quot;</span><span class="p">][</span><span class="s">&quot;LastUpdate&quot;</span><span class="p">]</span>
<span class="go">&#39;2008/05/24 06:45&#39;</span>
</pre></div>
</div>
<p>通过 <tt class="docutils literal"><span class="pre">record[&quot;DbInfo&quot;].keys()</span></tt> 可以获取存储在这个记录里面的其他信息。这里面最有用的信息之一是一个ESearch可用的
搜索值列表：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;DbInfo&quot;</span><span class="p">][</span><span class="s">&quot;FieldList&quot;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="s">&quot;</span><span class="si">%(Name)s</span><span class="s">, </span><span class="si">%(FullName)s</span><span class="s">, </span><span class="si">%(Description)s</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">field</span>
<span class="go">ALL, All Fields, All terms from all searchable fields</span>
<span class="go">UID, UID, Unique number assigned to publication</span>
<span class="go">FILT, Filter, Limits the records</span>
<span class="go">TITL, Title, Words in title of publication</span>
<span class="go">WORD, Text Word, Free text associated with publication</span>
<span class="go">MESH, MeSH Terms, Medical Subject Headings assigned to publication</span>
<span class="go">MAJR, MeSH Major Topic, MeSH terms of major importance to publication</span>
<span class="go">AUTH, Author, Author(s) of publication</span>
<span class="go">JOUR, Journal, Journal abbreviation of publication</span>
<span class="go">AFFL, Affiliation, Author&#39;s institutional affiliation and address</span>
<span class="gp">...</span>
</pre></div>
</div>
<p>这是一个很长的列表，但是间接的告诉你在使用PubMed的时候，你可以通过 <tt class="docutils literal"><span class="pre">Jones[AUTH]</span></tt> 搜索作者，或者通过
<tt class="docutils literal"><span class="pre">Sanger[AFFL]</span></tt> 将作者范围限制在Sanger Centre。这个会非常方便，特别是在你对某个数据库不太熟悉的时候。</p>
</div>
<div class="section" id="esearch-entrez">
<h2>9.3  ESearch: 搜索Entrez数据库<a class="headerlink" href="#esearch-entrez" title="Permalink to this headline">¶</a></h2>
<p>我们可以使用 <tt class="docutils literal"><span class="pre">Bio.Entrez.esearch()</span></tt> 来搜索任意的数据库。例如，我们在PubMed中搜索跟Biopython相关的文献：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>     <span class="c"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;pubmed&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s">&quot;biopython&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;IdList&quot;</span><span class="p">]</span>
<span class="go">[&#39;19304878&#39;, &#39;18606172&#39;, &#39;16403221&#39;, &#39;16377612&#39;, &#39;14871861&#39;, &#39;14630660&#39;, &#39;12230038&#39;]</span>
</pre></div>
</div>
<p>在输出的结果中，我们可以看到七个PubMed IDs（包括19304878，这个是Biopython应用笔记的PMID），你可以通过
EFetch来获取这些文献（请见章节 <a class="reference external" href="#sec:efetch">9.6</a> ）。</p>
<p>你也可以通过ESearch来搜索GenBank。我们将以在*Cypripedioideae* orchids中搜索*matK*基因为例，快速展示
一下（请见章节 <a class="reference external" href="#sec:entrez-einfo">9.2</a> 关于EInfo：一种查明你可以在哪个Entrez数据库中搜索的方法）。</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;nucleotide&quot;</span><span class="p">,</span><span class="n">term</span><span class="o">=</span><span class="s">&quot;Cypripedioideae[Orgn] AND matK[Gene]&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;Count&quot;</span><span class="p">]</span>
<span class="go">&#39;25&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;IdList&quot;</span><span class="p">]</span>
<span class="go">[&#39;126789333&#39;, &#39;37222967&#39;, &#39;37222966&#39;, &#39;37222965&#39;, ..., &#39;61585492&#39;]</span>
</pre></div>
</div>
<p>每个IDs(126789333, 37222967, 37222966, …)是GenBank的一个标识。请见Section <a class="reference external" href="#sec:efetch">9.6</a>
此章包含了怎样下载这些GenBank的记录的信息。</p>
<p>注意，不是像 <tt class="docutils literal"><span class="pre">Cypripedioideae[Orgn]</span></tt> 这样在搜索的时候加上特定的物种名字，而是需要在搜索的时候使用NCBI的
taxon ID，像 <tt class="docutils literal"><span class="pre">txid158330[Orgn]</span></tt> 这样。这个并没有记录在ESearch的帮助页面上，NCBI通过邮件回复解释了这个
问题。你可以通过经常和Entrez的网站接口互动，来推断搜索条目的格式。例如，在基因组搜索的时候加上 <tt class="docutils literal"><span class="pre">complete[prop]</span></tt>
可以把结果限制在完成的基因组上。</p>
<p>作为最后一个例子，让我们获取一个computational journal名字的列表：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;journals&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s">&quot;computational&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;Count&quot;</span><span class="p">]</span>
<span class="go">&#39;16&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;IdList&quot;</span><span class="p">]</span>
<span class="go">[&#39;30367&#39;, &#39;33843&#39;, &#39;33823&#39;, &#39;32989&#39;, &#39;33190&#39;, &#39;33009&#39;, &#39;31986&#39;,</span>
<span class="go"> &#39;34502&#39;, &#39;8799&#39;, &#39;22857&#39;, &#39;32675&#39;, &#39;20258&#39;, &#39;33859&#39;, &#39;32534&#39;,</span>
<span class="go"> &#39;32357&#39;, &#39;32249&#39;]</span>
</pre></div>
</div>
<p>同样，我们可以通过EFetch来获得关于每个journal IDs更多的消息。</p>
<p>ESearch有很多有用的参数——参见 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/esearch_help.html">ESearch 帮助页面</a>
来获取更多信息.</p>
</div>
<div class="section" id="epost-identifiers">
<h2>9.4  EPost: 上传identifiers的列表<a class="headerlink" href="#epost-identifiers" title="Permalink to this headline">¶</a></h2>
<p>EPost上传在后续搜索中将会用到的IDs的列表，参见 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/epost_help.html">EPost 帮助页面</a>
来获取更多信息. 通过 <tt class="docutils literal"><span class="pre">Bio.Entrez.epost()</span></tt> 函数可以在Biopython中实现。</p>
<p>为了举一个关于此用法的例子，假设你有一个想通过EFetch下载的IDs的长长的列表（可能是序列，也有可能是引用的
其他内容）。当你通过EFetch发出下载请求的时候，你的IDs列表、数据库等，将会被转变成一个长的URL，然后被发送
到服务器。如果IDs列表很长，URL也会很长，长的URL可能会断掉（比如，一些代理不能复制全部的内容）。</p>
<p>另外，你也可以把以上分成两步来完成，首先用EPost来上传IDs的列表（这个使用了一个内部的 “HTML post” ，而不是
“HTML get” ， 避开了long URL可能产生的问题）。由于历史记录的支持，你可以使用EFetch来指向这个长的IDs列表，
并且下载相关的数据。</p>
<p>让我们通过下面一个简单的例子来看看EPost是如何工作的——上传了一些PubMed的IDs：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>     <span class="c"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">id_list</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;19304878&quot;</span><span class="p">,</span> <span class="s">&quot;18606172&quot;</span><span class="p">,</span> <span class="s">&quot;16403221&quot;</span><span class="p">,</span> <span class="s">&quot;16377612&quot;</span><span class="p">,</span> <span class="s">&quot;14871861&quot;</span><span class="p">,</span> <span class="s">&quot;14630660&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">epost</span><span class="p">(</span><span class="s">&quot;pubmed&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">id_list</span><span class="p">))</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="go">&lt;?xml version=&quot;1.0&quot;?&gt;</span>
<span class="go">&lt;!DOCTYPE ePostResult PUBLIC &quot;-//NLM//DTD ePostResult, 11 May 2002//EN&quot;</span>
<span class="go"> &quot;http://www.ncbi.nlm.nih.gov/entrez/query/DTD/ePost_020511.dtd&quot;&gt;</span>
<span class="go">&lt;ePostResult&gt;</span>
<span class="go"> &lt;QueryKey&gt;1&lt;/QueryKey&gt;</span>
<span class="go"> &lt;WebEnv&gt;NCID_01_206841095_130.14.22.101_9001_1242061629&lt;/WebEnv&gt;</span>
<span class="go">&lt;/ePostResult&gt;</span>
</pre></div>
</div>
<p>返回的XML包含了两个重要的字符串， <tt class="docutils literal"><span class="pre">QueryKey</span></tt> 和 <tt class="docutils literal"><span class="pre">WebEnv</span></tt> ，两个字符串一起确定了之前的历史记录。你可以
使用其他的Entrez工具，例如EFetch，来提取这些值：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>     <span class="c"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">id_list</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;19304878&quot;</span><span class="p">,</span> <span class="s">&quot;18606172&quot;</span><span class="p">,</span> <span class="s">&quot;16403221&quot;</span><span class="p">,</span> <span class="s">&quot;16377612&quot;</span><span class="p">,</span> <span class="s">&quot;14871861&quot;</span><span class="p">,</span> <span class="s">&quot;14630660&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">search_results</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">Entrez</span><span class="o">.</span><span class="n">epost</span><span class="p">(</span><span class="s">&quot;pubmed&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">id_list</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">webenv</span> <span class="o">=</span> <span class="n">search_results</span><span class="p">[</span><span class="s">&quot;WebEnv&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query_key</span> <span class="o">=</span> <span class="n">search_results</span><span class="p">[</span><span class="s">&quot;QueryKey&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>第 <a class="reference external" href="#sec:entrez-webenv">9.15</a> 章节讲述了如何使用历史的特性。</p>
</div>
<div class="section" id="esummary-ids">
<h2>9.5  ESummary: 通过主要的IDs来获取摘要<a class="headerlink" href="#esummary-ids" title="Permalink to this headline">¶</a></h2>
<p>ESummary可以通过一个primary IDs来获取文章的摘要（参见 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/esummary_help.html">ESummary 帮助页面</a>
来获取更多信息）。在Biopython中，ESummary以 <tt class="docutils literal"><span class="pre">Bio.Entrez.esummary()</span></tt> 的形式出现。根据上面的搜索结果，
我们可以获得ID为30367杂志相关的更多信息：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>     <span class="c"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esummary</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;journals&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s">&quot;30367&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;Id&quot;</span><span class="p">]</span>
<span class="go">&#39;30367&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;Title&quot;</span><span class="p">]</span>
<span class="go">&#39;Computational biology and chemistry&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;Publisher&quot;</span><span class="p">]</span>
<span class="go">&#39;Pergamon,&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="efetch-entrez">
<h2>9.6  EFetch: 从Entrez下载更多的记录<a class="headerlink" href="#efetch-entrez" title="Permalink to this headline">¶</a></h2>
<p>当你想要从Entrez中提取完整的记录的时候，你可以使用EFetch。 在 <a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/entrez/query/static/efetch_help.html">EFetch的帮助页面</a>
可以查到EFetch可以起作用的数据库。</p>
<p>NCBI大部分的数据库都支持多种不同的文件格式。当使用 <tt class="docutils literal"><span class="pre">Bio.Entrez.efetch()</span></tt> 从Entrez下载特定的某种格式的时候，
需要 <tt class="docutils literal"><span class="pre">rettype</span></tt> 和或者 <tt class="docutils literal"><span class="pre">retmode</span></tt> 这些可选的参数。对于不同数据库类型不同的搭配在下面的网页中有描述：
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/efetch_help.html">NCBI efetch webpage</a>
(例如：
<a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/corehtml/query/static/efetchlit_help.html">literature</a>,
<a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/corehtml/query/static/efetchseq_help.html">sequences</a>
and
<a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/corehtml/query/static/efetchtax_help.html">taxonomy</a>).</p>
<p>一种常用的用法是下载FASTA或者GenBank/GenPept的文本格式 (接着可以使用 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 来解析, 参见 <a class="reference external" href="#sec:SeqIO_GenBank_Online">5.3.1</a>
和 <a class="reference external" href="#sec:efetch">9.6</a> ）。从上面 <em>Cypripedioideae</em> 的例子,我们可以通过 <tt class="docutils literal"><span class="pre">Bio.Entrez.efetch</span></tt>
从GenBank下载记录186972394。</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>     <span class="c"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s">&quot;186972394&quot;</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s">&quot;gb&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s">&quot;text&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">handle</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="go">LOCUS       EU490707                1302 bp    DNA     linear   PLN 05-MAY-2008</span>
<span class="go">DEFINITION  Selenipedium aequinoctiale maturase K (matK) gene, partial cds;</span>
<span class="go">            chloroplast.</span>
<span class="go">ACCESSION   EU490707</span>
<span class="go">VERSION     EU490707.1  GI:186972394</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE      chloroplast Selenipedium aequinoctiale</span>
<span class="go">  ORGANISM  Selenipedium aequinoctiale</span>
<span class="go">            Eukaryota; Viridiplantae; Streptophyta; Embryophyta; Tracheophyta;</span>
<span class="go">            Spermatophyta; Magnoliophyta; Liliopsida; Asparagales; Orchidaceae;</span>
<span class="go">            Cypripedioideae; Selenipedium.</span>
<span class="go">REFERENCE   1  (bases 1 to 1302)</span>
<span class="go">  AUTHORS   Neubig,K.M., Whitten,W.M., Carlsward,B.S., Blanco,M.A.,</span>
<span class="go">            Endara,C.L., Williams,N.H. and Moore,M.J.</span>
<span class="go">  TITLE     Phylogenetic utility of ycf1 in orchids</span>
<span class="go">  JOURNAL   Unpublished</span>
<span class="go">REFERENCE   2  (bases 1 to 1302)</span>
<span class="go">  AUTHORS   Neubig,K.M., Whitten,W.M., Carlsward,B.S., Blanco,M.A.,</span>
<span class="go">            Endara,C.L., Williams,N.H. and Moore,M.J.</span>
<span class="go">  TITLE     Direct Submission</span>
<span class="go">  JOURNAL   Submitted (14-FEB-2008) Department of Botany, University of</span>
<span class="go">            Florida, 220 Bartram Hall, Gainesville, FL 32611-8526, USA</span>
<span class="go">FEATURES             Location/Qualifiers</span>
<span class="go">     source          1..1302</span>
<span class="go">                     /organism=&quot;Selenipedium aequinoctiale&quot;</span>
<span class="go">                     /organelle=&quot;plastid:chloroplast&quot;</span>
<span class="go">                     /mol_type=&quot;genomic DNA&quot;</span>
<span class="go">                     /specimen_voucher=&quot;FLAS:Blanco 2475&quot;</span>
<span class="go">                     /db_xref=&quot;taxon:256374&quot;</span>
<span class="go">     gene            &lt;1..&gt;1302</span>
<span class="go">                     /gene=&quot;matK&quot;</span>
<span class="go">     CDS             &lt;1..&gt;1302</span>
<span class="go">                     /gene=&quot;matK&quot;</span>
<span class="go">                     /codon_start=1</span>
<span class="go">                     /transl_table=11</span>
<span class="go">                     /product=&quot;maturase K&quot;</span>
<span class="go">                     /protein_id=&quot;ACC99456.1&quot;</span>
<span class="go">                     /db_xref=&quot;GI:186972395&quot;</span>
<span class="go">                     /translation=&quot;IFYEPVEIFGYDNKSSLVLVKRLITRMYQQNFLISSVNDSNQKG</span>
<span class="go">                     FWGHKHFFSSHFSSQMVSEGFGVILEIPFSSQLVSSLEEKKIPKYQNLRSIHSIFPFL</span>
<span class="go">                     EDKFLHLNYVSDLLIPHPIHLEILVQILQCRIKDVPSLHLLRLLFHEYHNLNSLITSK</span>
<span class="go">                     KFIYAFSKRKKRFLWLLYNSYVYECEYLFQFLRKQSSYLRSTSSGVFLERTHLYVKIE</span>
<span class="go">                     HLLVVCCNSFQRILCFLKDPFMHYVRYQGKAILASKGTLILMKKWKFHLVNFWQSYFH</span>
<span class="go">                     FWSQPYRIHIKQLSNYSFSFLGYFSSVLENHLVVRNQMLENSFIINLLTKKFDTIAPV</span>
<span class="go">                     ISLIGSLSKAQFCTVLGHPISKPIWTDFSDSDILDRFCRICRNLCRYHSGSSKKQVLY</span>
<span class="go">                     RIKYILRLSCARTLARKHKSTVRTFMRRLGSGLLEEFFMEEE&quot;</span>
<span class="go">ORIGIN</span>
<span class="go">        1 attttttacg aacctgtgga aatttttggt tatgacaata aatctagttt agtacttgtg</span>
<span class="go">       61 aaacgtttaa ttactcgaat gtatcaacag aattttttga tttcttcggt taatgattct</span>
<span class="go">      121 aaccaaaaag gattttgggg gcacaagcat tttttttctt ctcatttttc ttctcaaatg</span>
<span class="go">      181 gtatcagaag gttttggagt cattctggaa attccattct cgtcgcaatt agtatcttct</span>
<span class="go">      241 cttgaagaaa aaaaaatacc aaaatatcag aatttacgat ctattcattc aatatttccc</span>
<span class="go">      301 tttttagaag acaaattttt acatttgaat tatgtgtcag atctactaat accccatccc</span>
<span class="go">      361 atccatctgg aaatcttggt tcaaatcctt caatgccgga tcaaggatgt tccttctttg</span>
<span class="go">      421 catttattgc gattgctttt ccacgaatat cataatttga atagtctcat tacttcaaag</span>
<span class="go">      481 aaattcattt acgccttttc aaaaagaaag aaaagattcc tttggttact atataattct</span>
<span class="go">      541 tatgtatatg aatgcgaata tctattccag tttcttcgta aacagtcttc ttatttacga</span>
<span class="go">      601 tcaacatctt ctggagtctt tcttgagcga acacatttat atgtaaaaat agaacatctt</span>
<span class="go">      661 ctagtagtgt gttgtaattc ttttcagagg atcctatgct ttctcaagga tcctttcatg</span>
<span class="go">      721 cattatgttc gatatcaagg aaaagcaatt ctggcttcaa agggaactct tattctgatg</span>
<span class="go">      781 aagaaatgga aatttcatct tgtgaatttt tggcaatctt attttcactt ttggtctcaa</span>
<span class="go">      841 ccgtatagga ttcatataaa gcaattatcc aactattcct tctcttttct ggggtatttt</span>
<span class="go">      901 tcaagtgtac tagaaaatca tttggtagta agaaatcaaa tgctagagaa ttcatttata</span>
<span class="go">      961 ataaatcttc tgactaagaa attcgatacc atagccccag ttatttctct tattggatca</span>
<span class="go">     1021 ttgtcgaaag ctcaattttg tactgtattg ggtcatccta ttagtaaacc gatctggacc</span>
<span class="go">     1081 gatttctcgg attctgatat tcttgatcga ttttgccgga tatgtagaaa tctttgtcgt</span>
<span class="go">     1141 tatcacagcg gatcctcaaa aaaacaggtt ttgtatcgta taaaatatat acttcgactt</span>
<span class="go">     1201 tcgtgtgcta gaactttggc acggaaacat aaaagtacag tacgcacttt tatgcgaaga</span>
<span class="go">     1261 ttaggttcgg gattattaga agaattcttt atggaagaag aa</span>
<span class="go">//</span>
</pre></div>
</div>
<p>参数 <tt class="docutils literal"><span class="pre">rettype=&quot;gb&quot;</span></tt> 和 <tt class="docutils literal"><span class="pre">retmode=&quot;text&quot;</span></tt> 让我们下载的数据为GenBank格式。</p>
<p>需要注意的是直到2009年，Entrez EFetch API要求使用 “genbank” 作为返回类型，然而现在NCBI坚持使用官方的
“gb” 或 “gbwithparts” （或者针对蛋白的“gp”) 返回类型。同样需要注意的是，直到2012年2月，
Entrez EFetch API默认的返回格式为纯文本格式文件，现在默认的为XML格式。</p>
<p>作为另外的选择，你也可以使用 <tt class="docutils literal"><span class="pre">rettype=&quot;fasta&quot;</span></tt> 来获取Fasta格式的文件；参见 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/efetchseq_help.html">EFetch Sequences 帮助页面</a> 。
记住，可选的数据格式决定于你要下载的数据库——请参见 <a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/entrez/query/static/efetch_help.html">EFetch 帮助页面</a>.</p>
<p>如果你要获取记录的格式是 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 所接受的一种格式(见章节 <a class="reference external" href="#chapter:Bio.SeqIO">5</a>),
你可以直接将其解析为一个 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> ：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span><span class="p">,</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s">&quot;186972394&quot;</span><span class="p">,</span><span class="n">rettype</span><span class="o">=</span><span class="s">&quot;gb&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s">&quot;text&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s">&quot;genbank&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">record</span>
<span class="go">ID: EU490707.1</span>
<span class="go">Name: EU490707</span>
<span class="go">Description: Selenipedium aequinoctiale maturase K (matK) gene, partial cds; chloroplast.</span>
<span class="go">Number of features: 3</span>
<span class="gp">...</span>
<span class="go">Seq(&#39;ATTTTTTACGAACCTGTGGAAATTTTTGGTTATGACAATAAATCTAGTTTAGTA...GAA&#39;, IUPACAmbiguousDNA())</span>
</pre></div>
</div>
<p>需要注意的是，一种更加典型的用法是先把序列数据保存到一个本地文件，<em>然后</em> 使用 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 来解析。这样就避免了
在运行脚本的时候需要重复的下载同样的文件，并减轻NCBI服务器的负载。例如：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>     <span class="c"># Always tell NCBI who you are</span>
<span class="n">filename</span> <span class="o">=</span> <span class="s">&quot;gi_186972394.gbk&quot;</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
    <span class="c"># Downloading...</span>
    <span class="n">net_handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;nucleotide&quot;</span><span class="p">,</span><span class="nb">id</span><span class="o">=</span><span class="s">&quot;186972394&quot;</span><span class="p">,</span><span class="n">rettype</span><span class="o">=</span><span class="s">&quot;gb&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s">&quot;text&quot;</span><span class="p">)</span>
    <span class="n">out_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span>
    <span class="n">out_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">net_handle</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
    <span class="n">out_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">net_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">print</span> <span class="s">&quot;Saved&quot;</span>

<span class="k">print</span> <span class="s">&quot;Parsing...&quot;</span>
<span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s">&quot;genbank&quot;</span><span class="p">)</span>
<span class="k">print</span> <span class="n">record</span>
</pre></div>
</div>
<p>为了得到XML格式的输出，你可以使用 <tt class="docutils literal"><span class="pre">Bio.Entrez.read()</span></tt> 函数和参数 <tt class="docutils literal"><span class="pre">retmode=&quot;xml&quot;</span></tt> 进行解析，：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s">&quot;186972394&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s">&quot;xml&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;GBSeq_definition&quot;</span><span class="p">]</span>
<span class="go">&#39;Selenipedium aequinoctiale maturase K (matK) gene, partial cds; chloroplast&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;GBSeq_source&quot;</span><span class="p">]</span>
<span class="go">&#39;chloroplast Selenipedium aequinoctiale&#39;</span>
</pre></div>
</div>
<p>就像这样处理数据。例如解析其他数据库特异的文件格式（例如，PubMed中用到的 <tt class="docutils literal"><span class="pre">MEDLINE</span></tt> 格式），请参见章节 <a class="reference external" href="#sec:entrez-specialized-parsers">9.12</a>.</p>
<p>如果你想使用 <tt class="docutils literal"><span class="pre">Bio.Entrez.esearch()</span></tt> 进行搜索，然后用 <tt class="docutils literal"><span class="pre">Bio.Entrez.efetch()</span></tt> 下载数据，那么你需要用到
WebEnv的历史特性，请参加见章节 <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
</div>
<div class="section" id="elink-ncbi-entrez">
<h2>9.7  ELink: 在NCBI Entrez中搜索相关的条目<a class="headerlink" href="#elink-ncbi-entrez" title="Permalink to this headline">¶</a></h2>
<p>ELink，在Biopython中是 <tt class="docutils literal"><span class="pre">Bio.Entrez.elink()</span></tt> ，可以用来在NCBI Entrez数据库中寻找相关的条目。例如，你
可以使用它在gene数据库中寻找核苷酸条目，或者其他很酷的事情。</p>
<p>让我们使用ELink来在2009年的 <em>Bioinformatics</em> 杂志中寻找与Biopython应用相关的文章。这篇文章的PubMed ID
是19304878：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pmid</span> <span class="o">=</span> <span class="s">&quot;19304878&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">Entrez</span><span class="o">.</span><span class="n">elink</span><span class="p">(</span><span class="n">dbfrom</span><span class="o">=</span><span class="s">&quot;pubmed&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">pmid</span><span class="p">))</span>
</pre></div>
</div>
<p>变量 <tt class="docutils literal"><span class="pre">record</span></tt> 包含了一个Python列表，列出了已经搜索过的数据库。因为我们特指了一个PubMed ID来搜索，所以
<tt class="docutils literal"><span class="pre">record</span></tt> 只包含了一个条目。这个条目是一个字典变量，包含了我们需要寻找的条目的信息，以及能搜索到的所有相关
的内容：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;DbFrom&quot;</span><span class="p">]</span>
<span class="go">&#39;pubmed&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;IdList&quot;</span><span class="p">]</span>
<span class="go">[&#39;19304878&#39;]</span>
</pre></div>
</div>
<p>键 <tt class="docutils literal"><span class="pre">&quot;LinkSetDb&quot;</span></tt> 包含了搜索结果，将每个目标数据库保存为一个列表。在我们这个搜索中，我们只在PubMed数据库
中找到了结果（尽管已经被分到了不同的分类）：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;LinkSetDb&quot;</span><span class="p">])</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">linksetdb</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;LinkSetDb&quot;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">linksetdb</span><span class="p">[</span><span class="s">&quot;DbTo&quot;</span><span class="p">],</span> <span class="n">linksetdb</span><span class="p">[</span><span class="s">&quot;LinkName&quot;</span><span class="p">],</span> <span class="nb">len</span><span class="p">(</span><span class="n">linksetdb</span><span class="p">[</span><span class="s">&quot;Link&quot;</span><span class="p">])</span>
<span class="gp">...</span>
<span class="go">pubmed pubmed_pubmed 110</span>
<span class="go">pubmed pubmed_pubmed_combined 6</span>
<span class="go">pubmed pubmed_pubmed_five 6</span>
<span class="go">pubmed pubmed_pubmed_reviews 5</span>
<span class="go">pubmed pubmed_pubmed_reviews_five 5</span>
</pre></div>
</div>
<p>实际的搜索结果被保存在键值为 <tt class="docutils literal"><span class="pre">&quot;Link&quot;</span></tt> 的字典下。在标准搜索下，总共找到了110个条目。让我们现在看看我们第一个
搜索结果：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;LinkSetDb&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;Link&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="go">{u&#39;Id&#39;: &#39;19304878&#39;}</span>
</pre></div>
</div>
<p>这个就是我们搜索的文章，从中并不能看到更多的结果，所以让我们来看看我们的第二个搜索结果：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;LinkSetDb&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;Link&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
<span class="go">{u&#39;Id&#39;: &#39;14630660&#39;}</span>
</pre></div>
</div>
<p>这个PubMed ID为14530660的文章是关于Biopython PDB解析器的。</p>
<p>我们通过一个循环来打印出所有的PubMed IDs：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">link</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;LinkSetDb&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;Link&quot;</span><span class="p">]</span> <span class="p">:</span> <span class="k">print</span> <span class="n">link</span><span class="p">[</span><span class="s">&quot;Id&quot;</span><span class="p">]</span>
<span class="go">19304878</span>
<span class="go">14630660</span>
<span class="go">18689808</span>
<span class="go">17121776</span>
<span class="go">16377612</span>
<span class="go">12368254</span>
<span class="go">......</span>
</pre></div>
</div>
<p>现在漂亮极了，但是对我个人而言，我对某篇文章是否被引用过更感兴趣。好吧，ELink也可以完成这个——至少对PubMed
Central的杂志来说是这样的（请见 Section <a class="reference external" href="#sec:elink-citations">9.15.3</a>）。</p>
<p>关于ELink的帮助，请见 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/elink_help.html">ELink 帮助页面</a> 。
这是一个关于 <a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/corehtml/query/static/entrezlinks.html">link names</a>
的整个的子页面， 描述了不同的数据库可以怎样交叉的索引。</p>
</div>
<div class="section" id="egquery">
<h2>9.8  EGQuery: 全局搜索- 统计搜索的条目<a class="headerlink" href="#egquery" title="Permalink to this headline">¶</a></h2>
<p>EGQuery提供搜索字段在每个Entrez数据库中的数目。当我们只需要知道在每个数据库中能找到的条目的个数，
而不需要知道具体搜索结果的时候，这个非常的有用（请见例子`9.14.2 &lt;#subsec:entrez_example_genbank&gt;`__ below）。</p>
<p>在这个例子中，我们使用 <tt class="docutils literal"><span class="pre">Bio.Entrez.egquery()</span></tt> 来获取跟 “Biopython” 相关的数目：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>     <span class="c"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">egquery</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s">&quot;biopython&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;eGQueryResult&quot;</span><span class="p">]:</span> <span class="k">print</span> <span class="n">row</span><span class="p">[</span><span class="s">&quot;DbName&quot;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s">&quot;Count&quot;</span><span class="p">]</span>
<span class="gp">...</span>
<span class="go">pubmed 6</span>
<span class="go">pmc 62</span>
<span class="go">journals 0</span>
<span class="gp">...</span>
</pre></div>
</div>
<p>请见 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/egquery_help.html">EGQuery 帮助页面</a>
获得更多信息.</p>
</div>
<div class="section" id="espell">
<h2>9.9  ESpell: 获得拼写建议<a class="headerlink" href="#espell" title="Permalink to this headline">¶</a></h2>
<p>ESpell可以检索拼写建议。在这个例子中，我们使用 <tt class="docutils literal"><span class="pre">Bio.Entrez.espell()</span></tt> 来获得Biopython正确的拼写：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>     <span class="c"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">espell</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s">&quot;biopythooon&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;Query&quot;</span><span class="p">]</span>
<span class="go">&#39;biopythooon&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;CorrectedQuery&quot;</span><span class="p">]</span>
<span class="go">&#39;biopython&#39;</span>
</pre></div>
</div>
<p>请见 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/espell_help.html">ESpell 帮助页面</a>
获得更多信息. 这个的主要用法是在使用GUI工具的时候为搜索的条目自动的提供拼写建议。</p>
</div>
<div class="section" id="entrez-xml">
<h2>9.10  解析大的Entrez XML文件<a class="headerlink" href="#entrez-xml" title="Permalink to this headline">¶</a></h2>
<p><tt class="docutils literal"><span class="pre">Entrez.read</span></tt> 函数将Entrez返回的结果读取到一个Python对象里面去，这个对象被保存在内存中。对于解析太大的
XML文件而内存不够时，可以使用 <tt class="docutils literal"><span class="pre">Entrez.parse</span></tt> 这个函数。这是一个生成器函数，它将一个一个的读取XML文件里面的内容。只有XML
文件是一个列表对象的时候，这个函数才有用（换句话说，如果在一个内存无限的计算机上 <tt class="docutils literal"><span class="pre">Entrez.read</span></tt> 将返回一个
Python列表）。</p>
<p>例如，你可以通过NCBI的FTP站点从Entrez Gene 数据库中下载某个物种全部的条目作为一个文件。这个文件可能很大。
作为一个例子，在2009年9月4日，文件 <tt class="docutils literal"><span class="pre">Homo_sapiens.ags.gz</span></tt> 包含了Entrez Gene数据库中人的序列，文件大小
有116576kB。这个文件是 <tt class="docutils literal"><span class="pre">ASN</span></tt> 格式，可以通过NCBI的 <tt class="docutils literal"><span class="pre">gene2xml</span></tt> 程序转成XML格式（请到NCBI的FTP站点获取
更多的信息）：</p>
<div class="code python highlight-python"><pre>gene2xml -b T -i Homo_sapiens.ags -o Homo_sapiens.xml</pre>
</div>
<p>XML结果文件有6.1GB. 在大多数电脑上尝试 <tt class="docutils literal"><span class="pre">Entrez.read</span></tt> 都会导致 <tt class="docutils literal"><span class="pre">MemoryError</span></tt> 。</p>
<p>XML文件 <tt class="docutils literal"><span class="pre">Homo_sapiens.xml</span></tt> 包含了一个Entrez gene记录的列表，每个对应于人的一个Entrez基因信息。 <tt class="docutils literal"><span class="pre">Entrez.parse</span></tt>
将一个一个的读取这些记录。这样你可以通过遍历每个记录的方式打印或者存储每个记录相关的信息。例如，下面这个脚本
遍历了Entrez基因里面的记录，打印了每个基因的数目和名字：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&quot;Homo_sapiens.xml&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>

<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">status</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s">&#39;Entrezgene_track-info&#39;</span><span class="p">][</span><span class="s">&#39;Gene-track&#39;</span><span class="p">][</span><span class="s">&#39;Gene-track_status&#39;</span><span class="p">]</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">status</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;value&#39;</span><span class="p">]</span><span class="o">==</span><span class="s">&#39;discontinued&#39;</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">continue</span>
<span class="gp">... </span>    <span class="n">geneid</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s">&#39;Entrezgene_track-info&#39;</span><span class="p">][</span><span class="s">&#39;Gene-track&#39;</span><span class="p">][</span><span class="s">&#39;Gene-track_geneid&#39;</span><span class="p">]</span>
<span class="gp">... </span>    <span class="n">genename</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s">&#39;Entrezgene_gene&#39;</span><span class="p">][</span><span class="s">&#39;Gene-ref&#39;</span><span class="p">][</span><span class="s">&#39;Gene-ref_locus&#39;</span><span class="p">]</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">geneid</span><span class="p">,</span> <span class="n">genename</span>
</pre></div>
</div>
<p>将会打印以下内容:</p>
<div class="code python highlight-python"><pre>1 A1BG
2 A2M
3 A2MP
8 AA
9 NAT1
10 NAT2
11 AACP
12 SERPINA3
13 AADAC
14 AAMP
15 AANAT
16 AARS
17 AAVS1
...</pre>
</div>
</div>
<div class="section" id="id1">
<h2>9.11  错误处理<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<p>当解析XML文件的时候，可能出现一下三个错误：</p>
<ul class="simple">
<li>这个文件可能不是以常规的 XML 文件格式开头；</li>
<li>这个文件可能不完整或者包含一些非 XML 格式的内容；</li>
<li>这个文件是正常的 XML 文件，但是包含和相关 DTD 文件无关的条目。</li>
</ul>
<p>第一种情况会在，例如，你尝试把一个 Fasta 文件当做 XML 文件来处理时发生：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&quot;NC_005816.fna&quot;</span><span class="p">)</span> <span class="c"># a Fasta file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/__init__.py&quot;</span>, line <span class="m">257</span>, in <span class="n">read</span>
    <span class="n">record</span> <span class="o">=</span> <span class="n">handler</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/Parser.py&quot;</span>, line <span class="m">164</span>, in <span class="n">read</span>
    <span class="k">raise</span> <span class="n">NotXMLError</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
<span class="gr">Bio.Entrez.Parser.NotXMLError</span>: <span class="n">Failed to parse the XML data (syntax error: line 1, column 0). Please make sure that the input data are in XML format.</span>
</pre></div>
</div>
<p>这时候，解析器找不到 <tt class="docutils literal"><span class="pre">&lt;?xml</span> <span class="pre">...</span></tt> 标签，而这是一个 XML 文件开始的标志，那么可以确定这个文件不是 XML 文件。</p>
<p>当你的文件是XML格式，但是是不完整的（例如，提前结束了），那么解析器会报CorruptedXMLError错误。下面
这个是一个XML文件提前结束的例子：</p>
<div class="code python highlight-python"><pre>&lt;?xml version="1.0"?&gt;
&lt;!DOCTYPE eInfoResult PUBLIC "-//NLM//DTD eInfoResult, 11 May 2002//EN" "http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd"&gt;
&lt;eInfoResult&gt;
&lt;DbList&gt;
        &lt;DbName&gt;pubmed&lt;/DbName&gt;
        &lt;DbName&gt;protein&lt;/DbName&gt;
        &lt;DbName&gt;nucleotide&lt;/DbName&gt;
        &lt;DbName&gt;nuccore&lt;/DbName&gt;
        &lt;DbName&gt;nucgss&lt;/DbName&gt;
        &lt;DbName&gt;nucest&lt;/DbName&gt;
        &lt;DbName&gt;structure&lt;/DbName&gt;
        &lt;DbName&gt;genome&lt;/DbName&gt;
        &lt;DbName&gt;books&lt;/DbName&gt;
        &lt;DbName&gt;cancerchromosomes&lt;/DbName&gt;
        &lt;DbName&gt;cdd&lt;/DbName&gt;</pre>
</div>
<p>这个会生成以下的日志文件：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/__init__.py&quot;</span>, line <span class="m">257</span>, in <span class="n">read</span>
    <span class="n">record</span> <span class="o">=</span> <span class="n">handler</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/Parser.py&quot;</span>, line <span class="m">160</span>, in <span class="n">read</span>
    <span class="k">raise</span> <span class="n">CorruptedXMLError</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
<span class="gr">Bio.Entrez.Parser.CorruptedXMLError</span>: <span class="n">Failed to parse the XML data (no element found: line 16, column 0). Please make sure that the input data are not corrupted.</span>

<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<p>注意，报错信息告诉你在XML文件的什么位置检测到了错误。</p>
<p>如果XML文件当中包含有对应DTD文件中没有描述的标签的时候，会发生第三类错误。以下是这样一个XML文件的例子：</p>
<div class="code python highlight-python"><pre>&lt;?xml version="1.0"?&gt;
&lt;!DOCTYPE eInfoResult PUBLIC "-//NLM//DTD eInfoResult, 11 May 2002//EN" "http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd"&gt;
&lt;eInfoResult&gt;
        &lt;DbInfo&gt;
        &lt;DbName&gt;pubmed&lt;/DbName&gt;
        &lt;MenuName&gt;PubMed&lt;/MenuName&gt;
        &lt;Description&gt;PubMed bibliographic record&lt;/Description&gt;
        &lt;Count&gt;20161961&lt;/Count&gt;
        &lt;LastUpdate&gt;2010/09/10 04:52&lt;/LastUpdate&gt;
        &lt;FieldList&gt;
                &lt;Field&gt;
...
                &lt;/Field&gt;
        &lt;/FieldList&gt;
        &lt;DocsumList&gt;
                &lt;Docsum&gt;
                        &lt;DsName&gt;PubDate&lt;/DsName&gt;
                        &lt;DsType&gt;4&lt;/DsType&gt;
                        &lt;DsTypeName&gt;string&lt;/DsTypeName&gt;
                &lt;/Docsum&gt;
                &lt;Docsum&gt;
                        &lt;DsName&gt;EPubDate&lt;/DsName&gt;
...
        &lt;/DbInfo&gt;
&lt;/eInfoResult&gt;</pre>
</div>
<p>在这个文件里面，因为一些原因，<tt class="docutils literal"><span class="pre">&lt;DocsumList&gt;</span></tt> （还有一些其他的）标签没有在DTD文件 <tt class="docutils literal"><span class="pre">eInfo_020511.dtd</span></tt>
中列出来，XML文件对应DTD文件的第二行会特别的描述出来。默认情况下，如果没有找到DTD文件中的标签，解析器
会中止并报ValidationError错误。</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&quot;einfo3.xml&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/__init__.py&quot;</span>, line <span class="m">257</span>, in <span class="n">read</span>
    <span class="n">record</span> <span class="o">=</span> <span class="n">handler</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/Parser.py&quot;</span>, line <span class="m">154</span>, in <span class="n">read</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">parser</span><span class="o">.</span><span class="n">ParseFile</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
  File <span class="nb">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/Parser.py&quot;</span>, line <span class="m">246</span>, in <span class="n">startElementHandler</span>
    <span class="k">raise</span> <span class="n">ValidationError</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
<span class="gr">Bio.Entrez.Parser.ValidationError</span>: <span class="n">Failed to find tag &#39;DocsumList&#39; in the DTD. To skip all tags that are not represented in the DTD, please call Bio.Entrez.read or Bio.Entrez.parse with validate=False.</span>
</pre></div>
</div>
<p>可选地，你可以让解析器跳过这样的标签，而不是报ValidationError错误。通过调用 <tt class="docutils literal"><span class="pre">Entrez.read</span></tt> 或者
<tt class="docutils literal"><span class="pre">Entrez.parse</span></tt> 并使参数 <tt class="docutils literal"><span class="pre">validate</span></tt> 等于False可以实现这个功能：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&quot;einfo3.xml&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span><span class="n">validate</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>
<span class="go">&gt;&gt;&gt;</span>
</pre></div>
</div>
<p>当然，XML文件中的tag没有出现在对应DTD文件中的信息，将不会在 <tt class="docutils literal"><span class="pre">Entrez.read</span></tt> 的返回记录中出现。</p>
</div>
<div class="section" id="id2">
<h2>9.12  专用的解析器<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h2>
<p>函数 <tt class="docutils literal"><span class="pre">Bio.Entrez.read()</span></tt> 可以处理大部分（如果不是所有的话）Entrez返回的XML文件。Entrez也可以
允许你通过其他格式来获取数据，有时候，这种方式在可读性上比XML文件格式更具优势（或者下载文件的大小）。</p>
<p>为了使用 <tt class="docutils literal"><span class="pre">Bio.Entrez.efetch()</span></tt> 函数从Entrez中提取一种特有的文件格式，需要指明 <tt class="docutils literal"><span class="pre">rettype</span></tt> 和或者或 <tt class="docutils literal"><span class="pre">retmode</span></tt>
等可选参数。不同的组合在 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/efetch_help.html">NCBI efetch的页面</a> 。
有对不同数据库的描述。</p>
<p>一个显然的例子是，你可能更想以FASTA或者 GenBank/GenPept ( 这些可以通过 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 来处理, 请见 Sections <a class="reference external" href="#sec:SeqIO_GenBank_Online">5.3.1</a>
和 <a class="reference external" href="#sec:efetch">9.6</a> ） 纯文本形式下载序列。对于文献数据库，Biopython包含了一个处理PubMed中
使用的 <tt class="docutils literal"><span class="pre">MEDLINE</span></tt> 格式的解析器。</p>
<div class="section" id="medline">
<h3>9.12.1  解析Medline记录<a class="headerlink" href="#medline" title="Permalink to this headline">¶</a></h3>
<p>你可以在 <tt class="docutils literal"><span class="pre">Bio.Medline</span></tt> 中找到Medline的解析器。假设你想处理包含一个Medline记录的 <tt class="docutils literal"><span class="pre">pubmed_result1.txt</span></tt>
文件。你可以在Biopython的 <tt class="docutils literal"><span class="pre">Tests\Medline</span></tt> 目录下找到这个文件，这个文件内容如下所示：</p>
<div class="code python highlight-python"><pre>PMID- 12230038
OWN - NLM
STAT- MEDLINE
DA  - 20020916
DCOM- 20030606
LR  - 20041117
PUBM- Print
IS  - 1467-5463 (Print)
VI  - 3
IP  - 3
DP  - 2002 Sep
TI  - The Bio* toolkits--a brief overview.
PG  - 296-302
AB  - Bioinformatics research is often difficult to do with commercial software. The
      Open Source BioPerl, BioPython and Biojava projects provide toolkits with
...</pre>
</div>
<p>我们首先打开文件，然后解析它：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Medline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">input</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&quot;pubmed_result1.txt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Medline</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="nb">input</span><span class="p">)</span>
</pre></div>
</div>
<p>现在 <tt class="docutils literal"><span class="pre">record</span></tt> 将 Medline记录以Python字典的形式保存起来：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;PMID&quot;</span><span class="p">]</span>
<span class="go">&#39;12230038&#39;</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;AB&quot;</span><span class="p">]</span>
<span class="go">&#39;Bioinformatics research is often difficult to do with commercial software.</span>
<span class="go">The Open Source BioPerl, BioPython and Biojava projects provide toolkits with</span>
<span class="go">multiple functionality that make it easier to create customised pipelines or</span>
<span class="go">analysis. This review briefly compares the quirks of the underlying languages</span>
<span class="go">and the functionality, documentation, utility and relative advantages of the</span>
<span class="go">Bio counterparts, particularly from the point of view of the beginning</span>
<span class="go">biologist programmer.&#39;</span>
</pre></div>
</div>
<p>用于Medline记录的键值可以相当模糊，使用</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">help</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
</pre></div>
</div>
<p>可以做一个简单的总结。</p>
<p>为了解析包含多个Medline记录的文件，你可以使用 <tt class="docutils literal"><span class="pre">parse</span></tt> 函数来代替：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Medline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">input</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&quot;pubmed_result2.txt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Medline</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="nb">input</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;TI&quot;</span><span class="p">]</span>
<span class="go">A high level interface to SCOP and ASTRAL implemented in python.</span>
<span class="go">GenomeDiagram: a python package for the visualization of large-scale genomic data.</span>
<span class="go">Open source clustering software.</span>
<span class="go">PDB file parser and structure class implemented in Python.</span>
</pre></div>
</div>
<p>你可以通过 <tt class="docutils literal"><span class="pre">Bio.Entrez.efetch</span></tt> 来下载Medline记录，而不是保存在某个文件里。例如，让我们来查看PubMed
里面跟Biopython相关的所有所有Medline记录：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>     <span class="c"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;pubmed&quot;</span><span class="p">,</span><span class="n">term</span><span class="o">=</span><span class="s">&quot;biopython&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;IdList&quot;</span><span class="p">]</span>
<span class="go">[&#39;19304878&#39;, &#39;18606172&#39;, &#39;16403221&#39;, &#39;16377612&#39;, &#39;14871861&#39;, &#39;14630660&#39;, &#39;12230038&#39;]</span>
</pre></div>
</div>
<p>现在我们使用 <tt class="docutils literal"><span class="pre">Bio.Entrez.efetch</span></tt> 来下载这些Medline记录:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">idlist</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;IdList&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;pubmed&quot;</span><span class="p">,</span><span class="nb">id</span><span class="o">=</span><span class="n">idlist</span><span class="p">,</span><span class="n">rettype</span><span class="o">=</span><span class="s">&quot;medline&quot;</span><span class="p">,</span><span class="n">retmode</span><span class="o">=</span><span class="s">&quot;text&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>这里，我们使 <tt class="docutils literal"><span class="pre">rettype=&quot;medline&quot;,</span> <span class="pre">retmode=&quot;text&quot;</span></tt> 来以纯文本形式的Medline格式来得到这些记录。现在
我们使用 <tt class="docutils literal"><span class="pre">Bio.Medline</span></tt> 来解析这些记录：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Medline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Medline</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;AU&quot;</span><span class="p">]</span>
<span class="go">[&#39;Cock PJ&#39;, &#39;Antao T&#39;, &#39;Chang JT&#39;, &#39;Chapman BA&#39;, &#39;Cox CJ&#39;, &#39;Dalke A&#39;, ..., &#39;de Hoon MJ&#39;]</span>
<span class="go">[&#39;Munteanu CR&#39;, &#39;Gonzalez-Diaz H&#39;, &#39;Magalhaes AL&#39;]</span>
<span class="go">[&#39;Casbon JA&#39;, &#39;Crooks GE&#39;, &#39;Saqi MA&#39;]</span>
<span class="go">[&#39;Pritchard L&#39;, &#39;White JA&#39;, &#39;Birch PR&#39;, &#39;Toth IK&#39;]</span>
<span class="go">[&#39;de Hoon MJ&#39;, &#39;Imoto S&#39;, &#39;Nolan J&#39;, &#39;Miyano S&#39;]</span>
<span class="go">[&#39;Hamelryck T&#39;, &#39;Manderick B&#39;]</span>
<span class="go">[&#39;Mangalam H&#39;]</span>
</pre></div>
</div>
<p>为了比对，我们展示了一个XML格式的例子：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">idlist</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;IdList&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;pubmed&quot;</span><span class="p">,</span><span class="nb">id</span><span class="o">=</span><span class="n">idlist</span><span class="p">,</span><span class="n">rettype</span><span class="o">=</span><span class="s">&quot;medline&quot;</span><span class="p">,</span><span class="n">retmode</span><span class="o">=</span><span class="s">&quot;xml&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;MedlineCitation&quot;</span><span class="p">][</span><span class="s">&quot;Article&quot;</span><span class="p">][</span><span class="s">&quot;ArticleTitle&quot;</span><span class="p">]</span>
<span class="go">Biopython: freely available Python tools for computational molecular biology and</span>
<span class="go"> bioinformatics.</span>
<span class="go">Enzymes/non-enzymes classification model complexity based on composition, sequence,</span>
<span class="go"> 3D and topological indices.</span>
<span class="go">A high level interface to SCOP and ASTRAL implemented in python.</span>
<span class="go">GenomeDiagram: a python package for the visualization of large-scale genomic data.</span>
<span class="go">Open source clustering software.</span>
<span class="go">PDB file parser and structure class implemented in Python.</span>
<span class="go">The Bio* toolkits--a brief overview.</span>
</pre></div>
</div>
<p>需要注意的是，在上面这两个例子当中，为了简便我们混合使用了 ESearch 和 EFetch。在这种情形下，NCBI 希望你
使用他们的历史记录特性，在下面章节中会讲到Section <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
</div>
<div class="section" id="geo">
<h3>9.12.2  解析GEO记录<a class="headerlink" href="#geo" title="Permalink to this headline">¶</a></h3>
<p>GEO ( <a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/">Gene Expression Omnibus</a> ) 是高通量基因表达和杂交芯片
数据的数据库。 <tt class="docutils literal"><span class="pre">Bio.Geo</span></tt> 模块可以用来解析GEO格式的数据。</p>
<p>下面的代码展示了怎样将一个名称为 <tt class="docutils literal"><span class="pre">GSE16.txt</span></tt> 的GEO文件存进一个记录，并打印该记录：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Geo</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&quot;GSE16.txt&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Geo</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">record</span>
</pre></div>
</div>
<p>你可以使用 ESearch 来搜索 “gds” 数据库 (GEO 数据集) :</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span> <span class="c"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;gds&quot;</span><span class="p">,</span><span class="n">term</span><span class="o">=</span><span class="s">&quot;GSE16&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;Count&quot;</span><span class="p">]</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;IdList&quot;</span><span class="p">]</span>
<span class="go">[&#39;200000016&#39;, &#39;100000028&#39;]</span>
</pre></div>
</div>
<p>通过Entrez网站，UID “200000016” 是GDS16，其他的hit “100000028” 是相关的平台。不幸的是，在写
这份指南的时候，NCBI貌似还不支持通过Entrez下载GEO文件（不论XML文件，还是SOFT格式的文件）。</p>
<p>然而，可以相当直接的通过 FTP <a class="reference external" href="ftp://ftp.ncbi.nih.gov/pub/geo/">ftp://ftp.ncbi.nih.gov/pub/geo/</a> 来下载 GEO 文件。
在这个例子当中，你需要的文件应该是 <a class="reference external" href="ftp://ftp.ncbi.nih.gov/pub/geo/DATA/SOFT/by_series/GSE16/GSE16_family.soft.gz">ftp://ftp.ncbi.nih.gov/pub/geo/DATA/SOFT/by_series/GSE16/GSE16_family.soft.gz</a>
（一个压缩文件，参见Python的gzip 模块）。</p>
</div>
<div class="section" id="unigene">
<h3>9.12.3  解析UniGene记录<a class="headerlink" href="#unigene" title="Permalink to this headline">¶</a></h3>
<p>UniGene是NCBI的转录组数据库，每个UniGene记录展示了该转录本在某个特定物种中相关的基因。一个典型的UniGene
记录如下所示：</p>
<div class="code python highlight-python"><pre>ID          Hs.2
TITLE       N-acetyltransferase 2 (arylamine N-acetyltransferase)
GENE        NAT2
CYTOBAND    8p22
GENE_ID     10
LOCUSLINK   10
HOMOL       YES
EXPRESS      bone| connective tissue| intestine| liver| liver tumor| normal| soft tissue/muscle tissue tumor| adult
RESTR_EXPR   adult
CHROMOSOME  8
STS         ACC=PMC310725P3 UNISTS=272646
STS         ACC=WIAF-2120 UNISTS=44576
STS         ACC=G59899 UNISTS=137181
...
STS         ACC=GDB:187676 UNISTS=155563
PROTSIM     ORG=10090; PROTGI=6754794; PROTID=NP_035004.1; PCT=76.55; ALN=288
PROTSIM     ORG=9796; PROTGI=149742490; PROTID=XP_001487907.1; PCT=79.66; ALN=288
PROTSIM     ORG=9986; PROTGI=126722851; PROTID=NP_001075655.1; PCT=76.90; ALN=288
...
PROTSIM     ORG=9598; PROTGI=114619004; PROTID=XP_519631.2; PCT=98.28; ALN=288

SCOUNT      38
SEQUENCE    ACC=BC067218.1; NID=g45501306; PID=g45501307; SEQTYPE=mRNA
SEQUENCE    ACC=NM_000015.2; NID=g116295259; PID=g116295260; SEQTYPE=mRNA
SEQUENCE    ACC=D90042.1; NID=g219415; PID=g219416; SEQTYPE=mRNA
SEQUENCE    ACC=D90040.1; NID=g219411; PID=g219412; SEQTYPE=mRNA
SEQUENCE    ACC=BC015878.1; NID=g16198419; PID=g16198420; SEQTYPE=mRNA
SEQUENCE    ACC=CR407631.1; NID=g47115198; PID=g47115199; SEQTYPE=mRNA
SEQUENCE    ACC=BG569293.1; NID=g13576946; CLONE=IMAGE:4722596; END=5'; LID=6989; SEQTYPE=EST; TRACE=44157214
...
SEQUENCE    ACC=AU099534.1; NID=g13550663; CLONE=HSI08034; END=5'; LID=8800; SEQTYPE=EST
//</pre>
</div>
<p>这个记录展示了这个转录本（如 <tt class="docutils literal"><span class="pre">SEQUENCE</span></tt> 行展示）是来自人的NAT2基因，编码en N-acetyltransferase。
<tt class="docutils literal"><span class="pre">PROTSIM</span></tt> 显示的是和NAT2显著相似的蛋白质， <tt class="docutils literal"><span class="pre">STS</span></tt> 展示的是基因组当中的STS位点。</p>
<p>我们使用 <tt class="docutils literal"><span class="pre">Bio.UniGene</span></tt> 模块来解析UniGene文件：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">UniGene</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">input</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&quot;myunigenefile.data&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">UniGene</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="nb">input</span><span class="p">)</span>
</pre></div>
</div>
<p><tt class="docutils literal"><span class="pre">UniGene.read</span></tt> 返回的是一个包含一些和UniGene记录的字段相对应属性的Python对象。例如，</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">ID</span>
<span class="go">&quot;Hs.2&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">title</span>
<span class="go">&quot;N-acetyltransferase 2 (arylamine N-acetyltransferase)&quot;</span>
</pre></div>
</div>
<p><tt class="docutils literal"><span class="pre">EXPRESS</span></tt> 和 <tt class="docutils literal"><span class="pre">RESTR_EXPR</span></tt> 两行被存储为字符串的Python列表：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="p">[</span><span class="s">&#39;bone&#39;</span><span class="p">,</span> <span class="s">&#39;connective tissue&#39;</span><span class="p">,</span> <span class="s">&#39;intestine&#39;</span><span class="p">,</span> <span class="s">&#39;liver&#39;</span><span class="p">,</span> <span class="s">&#39;liver tumor&#39;</span><span class="p">,</span> <span class="s">&#39;normal&#39;</span><span class="p">,</span> <span class="s">&#39;soft tissue/muscle tissue tumor&#39;</span><span class="p">,</span> <span class="s">&#39;adult&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>跟 <tt class="docutils literal"><span class="pre">STS</span></tt> , <tt class="docutils literal"><span class="pre">PROTSIM</span></tt> , 和 <tt class="docutils literal"><span class="pre">SEQUENCE</span></tt> 相关的特有的对象被保存在如下键所对应的字典中：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">sts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">acc</span>
<span class="go">&#39;PMC310725P3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">sts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">unists</span>
<span class="go">&#39;272646&#39;</span>
</pre></div>
</div>
<p>和 <tt class="docutils literal"><span class="pre">PROTSIM</span></tt> 、 <tt class="docutils literal"><span class="pre">SEQUENCE</span></tt> 这两行相似。</p>
<p>我们使用 <tt class="docutils literal"><span class="pre">Bio.UniGene</span></tt> 中的 <tt class="docutils literal"><span class="pre">parse</span></tt> 函数来处理一个文件中包含多个UniGene记录的情况：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">UniGene</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">input</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&quot;unigenerecords.data&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">UniGene</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="nb">input</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="n">record</span><span class="o">.</span><span class="n">ID</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="id3">
<h2>9.13  使用代理<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h2>
<p>通常状况下，你不需要使用代理，但是如果你的网络有问题的时候，我们有以下应对方法。在内部， <tt class="docutils literal"><span class="pre">Bio.Entrez</span></tt> 使用
一个标准的 Python 库 <tt class="docutils literal"><span class="pre">urllib</span></tt> 来访问 NCBI的服务器。这个将检查叫做 <tt class="docutils literal"><span class="pre">http_proxy</span></tt> 的环境变量来自动配置简单
的代理服务。不幸的是，这个模块不支持需要认证的代理。</p>
<p>你可以选择设定环境变量 <tt class="docutils literal"><span class="pre">http_proxy</span></tt> 。同样，你可以在Python脚本开头的地方设置这个参数，例如：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">import</span> <span class="nn">os</span>
<span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s">&quot;http_proxy&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;http://proxyhost.example.com:8080&quot;</span>
</pre></div>
</div>
<p>参见 <a class="reference external" href="http://www.python.org/doc/lib/module-urllib.html">urllib
文档</a> 获得更多信息。</p>
</div>
<div class="section" id="id4">
<h2>9.14  实例<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h2>
<div class="section" id="pubmedmedline">
<h3>9.14.1  PubMed和Medline<a class="headerlink" href="#pubmedmedline" title="Permalink to this headline">¶</a></h3>
<p>如果你是在医药领域或者对人类的问题感兴趣（或者尽管并你不感兴趣，大多数情况下也适用!），PubMed(<a class="reference external" href="http://www.ncbi.nlm.nih.gov/PubMed/">http://www.ncbi.nlm.nih.gov/PubMed/</a>)
是一个包含了各方面的非常优秀的资源。像其他的一样，我们希望能够通过 Python 脚本从中抓取一些信息。</p>
<p>在这个例子当中，我们要查询PubMed当中所有跟Orchids相关的文章(see section <a class="reference external" href="#sec:orchids">2.3</a> for our motivation)。
我们首先看看有多少这样的文章：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>     <span class="c"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">egquery</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s">&quot;orchid&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;eGQueryResult&quot;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s">&quot;DbName&quot;</span><span class="p">]</span><span class="o">==</span><span class="s">&quot;pubmed&quot;</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">print</span> <span class="n">row</span><span class="p">[</span><span class="s">&quot;Count&quot;</span><span class="p">]</span>
<span class="go">463</span>
</pre></div>
</div>
<p>现在我们使用 <tt class="docutils literal"><span class="pre">Bio.Entrez.efetch</span></tt> 这个函数来下载这463篇文章的PubMed IDs：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;pubmed&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s">&quot;orchid&quot;</span><span class="p">,</span> <span class="n">retmax</span><span class="o">=</span><span class="mi">463</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">idlist</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;IdList&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">idlist</span>
</pre></div>
</div>
<p>返回值是一个Python列表，包含了所有和orchids相关文章的PubMed IDs：</p>
<div class="code python highlight-python"><pre>['18680603', '18665331', '18661158', '18627489', '18627452', '18612381',
'18594007', '18591784', '18589523', '18579475', '18575811', '18575690',
...</pre>
</div>
<p>这样我们就得到了这些信息，显然我们想要得到对应的Medline records和更多额外的信息。这里，我们将以纯文本的
形式下载和Medline records相关的信息，然后使用 <tt class="docutils literal"><span class="pre">Bio.Medline</span></tt> 模块来解析他们：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Medline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;pubmed&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">idlist</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s">&quot;medline&quot;</span><span class="p">,</span>
<span class="go">                           retmode=&quot;text&quot;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Medline</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
</pre></div>
</div>
<p>注意 - 我们完成了一次搜索和获取，NCBI更希望你在这种情况下使用他们的历史记录支持。请见章节 <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
<p>请记住 <tt class="docutils literal"><span class="pre">records</span></tt> 是一个迭代器，所以你只能访问这些records一次。如果你想保存这些records，你需要把他们转成列表：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">records</span><span class="p">)</span>
</pre></div>
</div>
<p>现在让我们迭代这些records，然后分别打印每一个record的信息：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="s">&quot;title:&quot;</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;TI&quot;</span><span class="p">,</span> <span class="s">&quot;?&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="s">&quot;authors:&quot;</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;AU&quot;</span><span class="p">,</span> <span class="s">&quot;?&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="k">print</span> <span class="s">&quot;source:&quot;</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&quot;SO&quot;</span><span class="p">,</span> <span class="s">&quot;?&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="k">print</span>
</pre></div>
</div>
<p>这个的输出结果是这样的:</p>
<div class="code python highlight-python"><pre>title: Sex pheromone mimicry in the early spider orchid (ophrys sphegodes):
patterns of hydrocarbons as the key mechanism for pollination by sexual
deception [In Process Citation]
authors: ['Schiestl FP', 'Ayasse M', 'Paulus HF', 'Lofstedt C', 'Hansson BS',
'Ibarra F', 'Francke W']
source: J Comp Physiol [A] 2000 Jun;186(6):567-74</pre>
</div>
<p>特别有意思的是作者的列表，作者的列表会作为一个标准的Python列表返回。这使得用标准的Python工具操作和搜索
变得简单。例如，我们可以像下面的代码这样循环读取所有条目来查找某个特定的作者：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">search_author</span> <span class="o">=</span> <span class="s">&quot;Waits T&quot;</span>

<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="ow">not</span> <span class="s">&quot;AU&quot;</span> <span class="ow">in</span> <span class="n">record</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">continue</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">search_author</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;AU&quot;</span><span class="p">]:</span>
<span class="gp">... </span>        <span class="k">print</span> <span class="s">&quot;Author </span><span class="si">%s</span><span class="s"> found: </span><span class="si">%s</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">search_author</span><span class="p">,</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;SO&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>希望这个章节可以让你知道Entrez和Medline借口的能力和便利性和怎样同时使用他们。</p>
</div>
<div class="section" id="id5">
<h3>9.14.2  搜索，下载，和解析Entrez核酸记录<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<p>这里我们将展示一个关于远程Entrez查询的简单例子。在 <a class="reference external" href="#sec:orchids">2.3</a> 章节，我们讲到了使用NCBI
的Entrez网站来搜索 NCBI 的核酸数据库来获得关于Cypripedioideae的信息。现在我们看看如何使用Python脚本
自动的处理。在这个例子当中，我们仅仅展示如何使用Entrez模块来连接，获取结果，解析他们。</p>
<p>首先，我们在下载这些结果之前，使用EGQuery来计算结果的数目。EGQuery 将会告诉我们在每个数据库中分别有多少
搜索结果，但在我们这个例子当中，我们只对核苷酸感兴趣：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>     <span class="c"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">egquery</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s">&quot;Cypripedioideae&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;eGQueryResult&quot;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s">&quot;DbName&quot;</span><span class="p">]</span><span class="o">==</span><span class="s">&quot;nuccore&quot;</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">print</span> <span class="n">row</span><span class="p">[</span><span class="s">&quot;Count&quot;</span><span class="p">]</span>
<span class="go">814</span>
</pre></div>
</div>
<p>所以，我们预期能找到814个 Entrez 核酸记录（这是我在2008年得到的结果；在未来这个结果应该会增加）。如果你得
到了高的不可思议的结果数目时，你可能得重新考虑是否需要下载所有的这些结果，下载是我们的下一步：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s">&quot;Cypripedioideae&quot;</span><span class="p">,</span> <span class="n">retmax</span><span class="o">=</span><span class="mi">814</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
</pre></div>
</div>
<p>在这里, <tt class="docutils literal"><span class="pre">record</span></tt> 是一个包含了搜索结果和一些辅助信息的Python字典。仅仅作为参考信息，让我们看看在这些字典当中
究竟存储了些什么内容：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">record</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">[u&#39;Count&#39;, u&#39;RetMax&#39;, u&#39;IdList&#39;, u&#39;TranslationSet&#39;, u&#39;RetStart&#39;, u&#39;QueryTranslation&#39;]</span>
</pre></div>
</div>
<p>首先, 让我们检查看看我们得到了多少个结果:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;Count&quot;</span><span class="p">]</span>
<span class="go">&#39;814&#39;</span>
</pre></div>
</div>
<p>这个结果是我们所期望的。这814个结果被存在了 <tt class="docutils literal"><span class="pre">record['IdList']</span></tt> 中:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="s">&quot;IdList&quot;</span><span class="p">])</span>
<span class="go">814</span>
</pre></div>
</div>
<p>让我们看看前五个结果:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;IdList&quot;</span><span class="p">][:</span><span class="mi">5</span><span class="p">]</span>
<span class="go">[&#39;187237168&#39;, &#39;187372713&#39;, &#39;187372690&#39;, &#39;187372688&#39;, &#39;187372686&#39;]</span>
</pre></div>
</div>
<p>我们可以使用 <tt class="docutils literal"><span class="pre">efetch</span></tt> 来下载这些结果. 尽管你可以一个一个的下载这些记录，但为了减少 NCBI 服务器的负载，最好呢还是
一次性的下载所有的结果。然而在这个情况下，你应该完美的使用在后面章节 <a class="reference external" href="#sec:entrez-webenv">9.15</a> 中会要讲到的历史记录特性。</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">idlist</span> <span class="o">=</span> <span class="s">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">record</span><span class="p">[</span><span class="s">&quot;IdList&quot;</span><span class="p">][:</span><span class="mi">5</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">idlist</span>
<span class="go">187237168,187372713,187372690,187372688,187372686</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">idlist</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s">&quot;xml&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">records</span><span class="p">)</span>
<span class="go">5</span>
</pre></div>
</div>
<p>每个这样的records对应一个GenBank record.</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">[u&#39;GBSeq_moltype&#39;, u&#39;GBSeq_source&#39;, u&#39;GBSeq_sequence&#39;,</span>
<span class="go"> u&#39;GBSeq_primary-accession&#39;, u&#39;GBSeq_definition&#39;, u&#39;GBSeq_accession-version&#39;,</span>
<span class="go"> u&#39;GBSeq_topology&#39;, u&#39;GBSeq_length&#39;, u&#39;GBSeq_feature-table&#39;,</span>
<span class="go"> u&#39;GBSeq_create-date&#39;, u&#39;GBSeq_other-seqids&#39;, u&#39;GBSeq_division&#39;,</span>
<span class="go"> u&#39;GBSeq_taxonomy&#39;, u&#39;GBSeq_references&#39;, u&#39;GBSeq_update-date&#39;,</span>
<span class="go"> u&#39;GBSeq_organism&#39;, u&#39;GBSeq_locus&#39;, u&#39;GBSeq_strandedness&#39;]</span>

<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;GBSeq_primary-accession&quot;</span><span class="p">]</span>
<span class="go">DQ110336</span>

<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;GBSeq_other-seqids&quot;</span><span class="p">]</span>
<span class="go">[&#39;gb|DQ110336.1|&#39;, &#39;gi|187237168&#39;]</span>

<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;GBSeq_definition&quot;</span><span class="p">]</span>
<span class="go">Cypripedium calceolus voucher Davis 03-03 A maturase (matR) gene, partial cds;</span>
<span class="go">mitochondrial</span>

<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;GBSeq_organism&quot;</span><span class="p">]</span>
<span class="go">Cypripedium calceolus</span>
</pre></div>
</div>
<p>你可以用这个来快速的开始搜索 —— 但是对于频繁的使用请见  <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
</div>
<div class="section" id="genbank-record">
<h3>9.14.3  搜索、下载和解析GenBank record<a class="headerlink" href="#genbank-record" title="Permalink to this headline">¶</a></h3>
<p>GenBank record 格式是保存序列信息、序列特征和其他相关信息非常普遍的一种方法。这种格式是从 NCBI 数据库
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/">http://www.ncbi.nlm.nih.gov/</a> 获取信息非常好的一种方式 .</p>
<p>在这个例子当中，我们将展示怎样去查询 NCBI 数据库，根据query提取记录，然后使用 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 解析他们 ——
在 <a class="reference external" href="#sec:SeqIO_GenBank_Online">5.3.1</a> 中提到过这些。简单起见，这个例子*不会*使用 WebEnv 历史记录特性
—— 请到 Section <a class="reference external" href="#sec:entrez-webenv">9.15</a> 查看。</p>
<p>首先，我们想要查询找出要获取的记录的ID。这里我们快速的检索我们最喜欢的一个物种 <em>Opuntia</em> (多刺的梨型仙人掌)。我们
可以做一个快速的检索来获得所有满足要求的GIs（GenBank标志符）。首先我们看看有多少个记录：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>     <span class="c"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">egquery</span><span class="p">(</span><span class="n">term</span><span class="o">=</span><span class="s">&quot;Opuntia AND rpl16&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;eGQueryResult&quot;</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s">&quot;DbName&quot;</span><span class="p">]</span><span class="o">==</span><span class="s">&quot;nuccore&quot;</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">print</span> <span class="n">row</span><span class="p">[</span><span class="s">&quot;Count&quot;</span><span class="p">]</span>
<span class="gp">...</span>
<span class="go">9</span>
</pre></div>
</div>
<p>现在我们下载GenBank identifiers的列表：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;nuccore&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s">&quot;Opuntia AND rpl16&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gi_list</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="s">&quot;IdList&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gi_list</span>
<span class="go">[&#39;57240072&#39;, &#39;57240071&#39;, &#39;6273287&#39;, &#39;6273291&#39;, &#39;6273290&#39;, &#39;6273289&#39;, &#39;6273286&#39;,</span>
<span class="go">&#39;6273285&#39;, &#39;6273284&#39;]</span>
</pre></div>
</div>
<p>现在我们使用这些GIs来下载GenBank records —— 注意在老的Biopython版本中，你必须将GI号用逗号隔开传递给Entrez，例如
在 Biopython 1.59中，你可以传递一个列表，下面的内容会为你做转换：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gi_str</span> <span class="o">=</span> <span class="s">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gi_list</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;nuccore&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">gi_str</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s">&quot;gb&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s">&quot;text&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>如果你想看原始的 GenBank 文件，你可以从这个句柄中读取并打印结果：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">text</span> <span class="o">=</span> <span class="n">handle</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">text</span>
<span class="go">LOCUS       AY851612                 892 bp    DNA     linear   PLN 10-APR-2007</span>
<span class="go">DEFINITION  Opuntia subulata rpl16 gene, intron; chloroplast.</span>
<span class="go">ACCESSION   AY851612</span>
<span class="go">VERSION     AY851612.1  GI:57240072</span>
<span class="go">KEYWORDS    .</span>
<span class="go">SOURCE      chloroplast Austrocylindropuntia subulata</span>
<span class="go">  ORGANISM  Austrocylindropuntia subulata</span>
<span class="go">            Eukaryota; Viridiplantae; Streptophyta; Embryophyta; Tracheophyta;</span>
<span class="go">            Spermatophyta; Magnoliophyta; eudicotyledons; core eudicotyledons;</span>
<span class="go">            Caryophyllales; Cactaceae; Opuntioideae; Austrocylindropuntia.</span>
<span class="go">REFERENCE   1  (bases 1 to 892)</span>
<span class="go">  AUTHORS   Butterworth,C.A. and Wallace,R.S.</span>
<span class="gp">...</span>
</pre></div>
</div>
<p>在这个例子当中，我们只是得到了原始的记录。为了得到对Python友好的格式，我们可以使用 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 将GenBank
数据转化成 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 对象，包括 <tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 对象 (请见 Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>):</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;nuccore&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="n">gi_str</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s">&quot;gb&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s">&quot;text&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">handle</span><span class="p">,</span> <span class="s">&quot;gb&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>我们现在可以逐个查看这些record来寻找我们感兴趣的信息：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="o">...</span>    <span class="k">print</span> <span class="s">&quot;</span><span class="si">%s</span><span class="s">, length </span><span class="si">%i</span><span class="s">, with </span><span class="si">%i</span><span class="s"> features&quot;</span> \
<span class="gp">&gt;&gt;&gt; </span><span class="o">...</span>           <span class="o">%</span> <span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">))</span>
<span class="go">AY851612, length 892, with 3 features</span>
<span class="go">AY851611, length 881, with 3 features</span>
<span class="go">AF191661, length 895, with 3 features</span>
<span class="go">AF191665, length 902, with 3 features</span>
<span class="go">AF191664, length 899, with 3 features</span>
<span class="go">AF191663, length 899, with 3 features</span>
<span class="go">AF191660, length 893, with 3 features</span>
<span class="go">AF191659, length 894, with 3 features</span>
<span class="go">AF191658, length 896, with 3 features</span>
</pre></div>
</div>
<p>使用这些自动的查询提取功能相对于手动处理是一个很大的进步。尽管这些模块需要遵守NCBI每秒钟最多三次的规则，然而NCBI
有其他像避开高峰时刻的建议。请见章节 <a class="reference external" href="#sec:entrez-guidelines">9.1</a>. 尤其需要注意的是，这个例子没有
用到 WebEnv 历史记录特性。你应该使用这个来完成一些琐碎的搜索和下载的工作，请见章节 <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
<p>最后，如果你计划重复你的分析，你应该下载这些record <em>一次</em> ，然后将他们保存在你的硬盘里，在本地进行分析；而不是
从 NCBI 下载之后就马上进行分析（像这个例子一样）。</p>
</div>
<div class="section" id="id6">
<h3>9.14.4  查看物种的谱系关系<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<p>仍然以植物为例子，让我们找出Cyripedioideae兰花家族的谱系。首先让我们在Taxonomy数据库中查找跟Cypripedioideae
相关的记录，确实找到了一个确切的 NCBI taxonomy 标识号：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>     <span class="c"># Always tell NCBI who you are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;Taxonomy&quot;</span><span class="p">,</span> <span class="n">term</span><span class="o">=</span><span class="s">&quot;Cypripedioideae&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;IdList&quot;</span><span class="p">]</span>
<span class="go">[&#39;158330&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="p">[</span><span class="s">&quot;IdList&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&#39;158330&#39;</span>
</pre></div>
</div>
<p>现在，我们使用 <tt class="docutils literal"><span class="pre">efetch</span></tt> 从 Taxonomy 数据库中下载这些条目，然后解析它：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;Taxonomy&quot;</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s">&quot;158330&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s">&quot;xml&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">records</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="p">)</span>
</pre></div>
</div>
<p>再次，这个record保存了许多的信息：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">[u&#39;Lineage&#39;, u&#39;Division&#39;, u&#39;ParentTaxId&#39;, u&#39;PubDate&#39;, u&#39;LineageEx&#39;,</span>
<span class="go"> u&#39;CreateDate&#39;, u&#39;TaxId&#39;, u&#39;Rank&#39;, u&#39;GeneticCode&#39;, u&#39;ScientificName&#39;,</span>
<span class="go"> u&#39;MitoGeneticCode&#39;, u&#39;UpdateDate&#39;]</span>
</pre></div>
</div>
<p>我们可以直接从这个record获得谱系信息：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">records</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;Lineage&quot;</span><span class="p">]</span>
<span class="go">&#39;cellular organisms; Eukaryota; Viridiplantae; Streptophyta; Streptophytina;</span>
<span class="go"> Embryophyta; Tracheophyta; Euphyllophyta; Spermatophyta; Magnoliophyta;</span>
<span class="go"> Liliopsida; Asparagales; Orchidaceae&#39;</span>
</pre></div>
</div>
<p>这个record数据包含的信息远远超过在这里显示的 —— 例如查看 <tt class="docutils literal"><span class="pre">&quot;LineageEx&quot;</span></tt> 而不是 <tt class="docutils literal"><span class="pre">&quot;Lineage&quot;</span></tt> 相关的
信息，你也可以得到谱系里面的 NCBI taxon 标识号信息。</p>
</div>
</div>
<div class="section" id="webenv">
<h2>9.15  使用历史记录和WebEnv<a class="headerlink" href="#webenv" title="Permalink to this headline">¶</a></h2>
<p>通常，你想做一系列相关的查询。最典型的是，进行一个搜索，精炼搜索，然后提取详细的搜索结果。你 <em>可以</em> 通过一系列
独立的调用Entrez来完成这些工作。然而，NCBI更希望你利用历史记录支持的优势来完成这个 - 例如将ESearch
和EFetch结合起来。</p>
<p>另外一个关于历史记录支持，典型的使用是结合EPost和EFetch。你可以使用EPost来上传一个标识号的
列表，这样就开始一些新的history session。接下来你就可以用EFetch指向这个session来下载这些数据（而不是那些
标识号）。</p>
<div class="section" id="history">
<h3>9.15.1  利用 history 来搜索和下载序列<a class="headerlink" href="#history" title="Permalink to this headline">¶</a></h3>
<p>假设我们想搜索和下载所有的 <em>Opuntia</em> rpl16核酸序列，然后将它们保存到一个FASTA文件里。就像章节 <a class="reference external" href="#sec:entrez-search-fetch-genbank">9.14.3</a>
里一样, 我们可以简单的用 <tt class="docutils literal"><span class="pre">Bio.Entrez.esearch()</span></tt> 得到一个GI号的列表，然后调用 <tt class="docutils literal"><span class="pre">Bio.Entrez.efetch()</span></tt>
来下载他们。</p>
<p>然而，被认同的方法是使用历史记录特性来进行搜索。然后，我们可以通过指向这些搜索结果的引用来获取他们 -
NCBI 将会提前进行缓冲。</p>
<p>为此，调用 <tt class="docutils literal"><span class="pre">Bio.Entrez.esearch()</span></tt> 是正常的，但是需要额外的 <tt class="docutils literal"><span class="pre">usehistory=&quot;y&quot;</span></tt> 参数，</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;history.user@example.com&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">search_handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;nucleotide&quot;</span><span class="p">,</span><span class="n">term</span><span class="o">=</span><span class="s">&quot;Opuntia[orgn] and rpl16&quot;</span><span class="p">,</span>
<span class="go">                                   usehistory=&quot;y&quot;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">search_results</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">search_handle</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">search_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>当你得到XML输出的时候，它仍然包括了常见的搜索结果：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gi_list</span> <span class="o">=</span> <span class="n">search_results</span><span class="p">[</span><span class="s">&quot;IdList&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">count</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">search_results</span><span class="p">[</span><span class="s">&quot;Count&quot;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="n">count</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">gi_list</span><span class="p">)</span>
</pre></div>
</div>
<p>然而，你将得到两个额外的信息， <tt class="docutils literal"><span class="pre">WebEnv</span></tt> 会话cookie 和 <tt class="docutils literal"><span class="pre">QueryKey</span></tt> :</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">webenv</span> <span class="o">=</span> <span class="n">search_results</span><span class="p">[</span><span class="s">&quot;WebEnv&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query_key</span> <span class="o">=</span> <span class="n">search_results</span><span class="p">[</span><span class="s">&quot;QueryKey&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>将这些值保存到 <tt class="docutils literal"><span class="pre">session_cookie</span></tt> 和 <tt class="docutils literal"><span class="pre">query_key</span></tt> 后，我们可以使用它们作为 <tt class="docutils literal"><span class="pre">Bio.Entrez.efetch()</span></tt> 的
参数，而不用提供GI numbers的identifiers。</p>
<p>对于小数据量你一次下载所有的数据也没有关系，但是最好能够分批下载。你可以使用 <tt class="docutils literal"><span class="pre">restart</span></tt> 和 <tt class="docutils literal"><span class="pre">retmax</span></tt> 来说明
哪一部分搜索结果是你想得到的（条目以0开始计算，返回结果的最大数目）。例如：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="n">batch_size</span> <span class="o">=</span> <span class="mi">3</span>
<span class="n">out_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&quot;orchid_rpl16.fasta&quot;</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">start</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">count</span><span class="p">,</span><span class="n">batch_size</span><span class="p">):</span>
    <span class="n">end</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">count</span><span class="p">,</span> <span class="n">start</span><span class="o">+</span><span class="n">batch_size</span><span class="p">)</span>
    <span class="k">print</span> <span class="s">&quot;Going to download record </span><span class="si">%i</span><span class="s"> to </span><span class="si">%i</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">start</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
    <span class="n">fetch_handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;nucleotide&quot;</span><span class="p">,</span> <span class="n">rettype</span><span class="o">=</span><span class="s">&quot;fasta&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s">&quot;text&quot;</span><span class="p">,</span>
                                 <span class="n">retstart</span><span class="o">=</span><span class="n">start</span><span class="p">,</span> <span class="n">retmax</span><span class="o">=</span><span class="n">batch_size</span><span class="p">,</span>
                                 <span class="n">webenv</span><span class="o">=</span><span class="n">webenv</span><span class="p">,</span> <span class="n">query_key</span><span class="o">=</span><span class="n">query_key</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">fetch_handle</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
    <span class="n">fetch_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">out_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">out_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>我们以此为例来说明，这个例子分三次来下载FASTA records。除非你是要下载基因组或者染色体数目，你最好选取一个
比较大的batch大小。</p>
</div>
<div class="section" id="id7">
<h3>9.15.2  利用history来搜索和下载综述<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<p>这是另外一个history的例子，搜索过去几年当中发表的关于 <em>Opuntia</em> 的文章，然后下载到一个MedLine格式的文件里：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;history.user@example.com&quot;</span>
<span class="n">search_results</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">Entrez</span><span class="o">.</span><span class="n">esearch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;pubmed&quot;</span><span class="p">,</span>
                                            <span class="n">term</span><span class="o">=</span><span class="s">&quot;Opuntia[ORGN]&quot;</span><span class="p">,</span>
                                            <span class="n">reldate</span><span class="o">=</span><span class="mi">365</span><span class="p">,</span> <span class="n">datetype</span><span class="o">=</span><span class="s">&quot;pdat&quot;</span><span class="p">,</span>
                                            <span class="n">usehistory</span><span class="o">=</span><span class="s">&quot;y&quot;</span><span class="p">))</span>
<span class="n">count</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">search_results</span><span class="p">[</span><span class="s">&quot;Count&quot;</span><span class="p">])</span>
<span class="k">print</span> <span class="s">&quot;Found </span><span class="si">%i</span><span class="s"> results&quot;</span> <span class="o">%</span> <span class="n">count</span>

<span class="n">batch_size</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">out_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s">&quot;recent_orchid_papers.txt&quot;</span><span class="p">,</span> <span class="s">&quot;w&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">start</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">count</span><span class="p">,</span><span class="n">batch_size</span><span class="p">):</span>
    <span class="n">end</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">count</span><span class="p">,</span> <span class="n">start</span><span class="o">+</span><span class="n">batch_size</span><span class="p">)</span>
    <span class="k">print</span> <span class="s">&quot;Going to download record </span><span class="si">%i</span><span class="s"> to </span><span class="si">%i</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">start</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
    <span class="n">fetch_handle</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">efetch</span><span class="p">(</span><span class="n">db</span><span class="o">=</span><span class="s">&quot;pubmed&quot;</span><span class="p">,</span>
                                 <span class="n">rettype</span><span class="o">=</span><span class="s">&quot;medline&quot;</span><span class="p">,</span> <span class="n">retmode</span><span class="o">=</span><span class="s">&quot;text&quot;</span><span class="p">,</span>
                                 <span class="n">retstart</span><span class="o">=</span><span class="n">start</span><span class="p">,</span> <span class="n">retmax</span><span class="o">=</span><span class="n">batch_size</span><span class="p">,</span>
                                 <span class="n">webenv</span><span class="o">=</span><span class="n">search_results</span><span class="p">[</span><span class="s">&quot;WebEnv&quot;</span><span class="p">],</span>
                                 <span class="n">query_key</span><span class="o">=</span><span class="n">search_results</span><span class="p">[</span><span class="s">&quot;QueryKey&quot;</span><span class="p">])</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">fetch_handle</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
    <span class="n">fetch_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">out_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="n">out_handle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
<p>在写这份文档的时候，这个搜索返回了28个匹配结果 - 但是因为这个是跟时间相关的搜索，因此返回结果会发生变化。
像在上面 <a class="reference external" href="#subsec:entrez-and-medline">9.12.1</a> 讲到的一样, 你可以使用 <tt class="docutils literal"><span class="pre">Bio.Medline</span></tt> 来解析
保存下来的记录。</p>
</div>
<div class="section" id="id8">
<h3>9.15.3  搜索引用文章<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<p>回到 Section <a class="reference external" href="#sec:elink">9.7</a> 我们提到可以使用ELink来搜索制定文章的引用。不幸的是，这个只包含
PubMed Central（为PubMed中所有文献来做这个事情，意味这NIH将要付出更多的工作）包含的那些杂志。让我们以
Biopython PDB parser文章为例来试试看， PubMed ID 14630660：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">Entrez</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Entrez</span><span class="o">.</span><span class="n">email</span> <span class="o">=</span> <span class="s">&quot;A.N.Other@example.com&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pmid</span> <span class="o">=</span> <span class="s">&quot;14630660&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">Entrez</span><span class="o">.</span><span class="n">elink</span><span class="p">(</span><span class="n">dbfrom</span><span class="o">=</span><span class="s">&quot;pubmed&quot;</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="s">&quot;pmc&quot;</span><span class="p">,</span>
<span class="gp">... </span>                                   <span class="n">LinkName</span><span class="o">=</span><span class="s">&quot;pubmed_pmc_refs&quot;</span><span class="p">,</span> <span class="n">from_uid</span><span class="o">=</span><span class="n">pmid</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pmc_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">link</span><span class="p">[</span><span class="s">&quot;Id&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">link</span> <span class="ow">in</span> <span class="n">results</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;LinkSetDb&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;Link&quot;</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pmc_ids</span>
<span class="go">[&#39;2744707&#39;, &#39;2705363&#39;, &#39;2682512&#39;, ..., &#39;1190160&#39;]</span>
</pre></div>
</div>
<p>好极了 - 11篇文章。但是为什么没有Biopython应用笔记（PubMed ID 19304878）呢？好吧，你可能已经从变量的
名称中猜到了，实际上他们不是PubMed IDs，而是PubMed Central IDs。我们的应用笔记是列表当中第三个引用的
文章， PMCID 2682512。</p>
<p>那么，如果（像我）你希望得到的是PubMed IDs的列表的话，该怎么做呢？好吧，你可以使用再次使用ELink来更改他们。
这将成为两步处理，所以你应该使用历史记录特性来完成这个工作（章节 <a class="reference external" href="#sec:entrez-webenv">9.15</a>）。</p>
<p>但是首先，让我们使用更直接的方法来进行第二次调用ELink：</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">results2</span> <span class="o">=</span> <span class="n">Entrez</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">Entrez</span><span class="o">.</span><span class="n">elink</span><span class="p">(</span><span class="n">dbfrom</span><span class="o">=</span><span class="s">&quot;pmc&quot;</span><span class="p">,</span> <span class="n">db</span><span class="o">=</span><span class="s">&quot;pubmed&quot;</span><span class="p">,</span> <span class="n">LinkName</span><span class="o">=</span><span class="s">&quot;pmc_pubmed&quot;</span><span class="p">,</span>
<span class="gp">... </span>                                    <span class="n">from_uid</span><span class="o">=</span><span class="s">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pmc_ids</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pubmed_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">link</span><span class="p">[</span><span class="s">&quot;Id&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">link</span> <span class="ow">in</span> <span class="n">results2</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;LinkSetDb&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s">&quot;Link&quot;</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pubmed_ids</span>
<span class="go">[&#39;19698094&#39;, &#39;19450287&#39;, &#39;19304878&#39;, ..., &#39;15985178&#39;]</span>
</pre></div>
</div>
<p>这次，你可以立即的看到Biopython应用笔记作为第三个hit（PubMed ID 19304878）。</p>
<p>现在，让我们重新使用历史记录再试一遍 … <em>TODO</em>.</p>
<p>最终，不要忘记在Entrez调用的时候，加上你 <em>自己</em> 的电子邮箱地址。</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">第9章  访问NCBI Entrez数据库</a><ul>
<li><a class="reference internal" href="#entrez">9.1  Entrez 简介</a></li>
<li><a class="reference internal" href="#einfo-entrez">9.2  EInfo: 获取Entrez数据库的信息</a></li>
<li><a class="reference internal" href="#esearch-entrez">9.3  ESearch: 搜索Entrez数据库</a></li>
<li><a class="reference internal" href="#epost-identifiers">9.4  EPost: 上传identifiers的列表</a></li>
<li><a class="reference internal" href="#esummary-ids">9.5  ESummary: 通过主要的IDs来获取摘要</a></li>
<li><a class="reference internal" href="#efetch-entrez">9.6  EFetch: 从Entrez下载更多的记录</a></li>
<li><a class="reference internal" href="#elink-ncbi-entrez">9.7  ELink: 在NCBI Entrez中搜索相关的条目</a></li>
<li><a class="reference internal" href="#egquery">9.8  EGQuery: 全局搜索- 统计搜索的条目</a></li>
<li><a class="reference internal" href="#espell">9.9  ESpell: 获得拼写建议</a></li>
<li><a class="reference internal" href="#entrez-xml">9.10  解析大的Entrez XML文件</a></li>
<li><a class="reference internal" href="#id1">9.11  错误处理</a></li>
<li><a class="reference internal" href="#id2">9.12  专用的解析器</a><ul>
<li><a class="reference internal" href="#medline">9.12.1  解析Medline记录</a></li>
<li><a class="reference internal" href="#geo">9.12.2  解析GEO记录</a></li>
<li><a class="reference internal" href="#unigene">9.12.3  解析UniGene记录</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id3">9.13  使用代理</a></li>
<li><a class="reference internal" href="#id4">9.14  实例</a><ul>
<li><a class="reference internal" href="#pubmedmedline">9.14.1  PubMed和Medline</a></li>
<li><a class="reference internal" href="#id5">9.14.2  搜索，下载，和解析Entrez核酸记录</a></li>
<li><a class="reference internal" href="#genbank-record">9.14.3  搜索、下载和解析GenBank record</a></li>
<li><a class="reference internal" href="#id6">9.14.4  查看物种的谱系关系</a></li>
</ul>
</li>
<li><a class="reference internal" href="#webenv">9.15  使用历史记录和WebEnv</a><ul>
<li><a class="reference internal" href="#history">9.15.1  利用 history 来搜索和下载序列</a></li>
<li><a class="reference internal" href="#id7">9.15.2  利用history来搜索和下载综述</a></li>
<li><a class="reference internal" href="#id8">9.15.3  搜索引用文章</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="chr08.html"
                        title="previous chapter">第8章  BLAST和其他序列搜索工具(<em>实验性质的代码</em>)</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="chr10.html"
                        title="next chapter">第10章 Swiss-Prot和ExPASy</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/chr09.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="chr10.html" title="第10章 Swiss-Prot和ExPASy"
             >next</a> |</li>
        <li class="right" >
          <a href="chr08.html" title="第8章 BLAST和其他序列搜索工具(实验性质的代码)"
             >previous</a> |</li>
        <li><a href="index.html">biopython_cn 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, biopythoners.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.
    </div>
  </body>
</html>