<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>第4章 序列注释对象 &mdash; biopython_cn 0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="biopython_cn 0.1 documentation" href="index.html" />
    <link rel="next" title="第5章 序列输入和输出" href="chr05.html" />
    <link rel="prev" title="第3章 生物序列对象" href="chr03.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="chr05.html" title="第5章 序列输入和输出"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="chr03.html" title="第3章 生物序列对象"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">biopython_cn 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="id1">
<h1>第4章  序列注释对象<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h1>
<p>Chapter <a class="reference external" href="#chapter:Bio.Seq">3</a> 介绍了序列对象的基本情况。紧接上章的 <tt class="docutils literal"><span class="pre">Seq</span></tt> 类，这章主要讲Sequence record 或称之为 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 类, 该类在 <tt class="docutils literal"><span class="pre">Bio.SeqRecord</span></tt> 模块中有定义。 它（见 <tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 对象）可使序列与高级属性（如identifiers 和 features）关联。其应用贯穿序列输入/输出的交互界面 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 过程中 （详见Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> ）。</p>
<p>如读者只需处理FASTA格式的序列文件等简单数据,可略过本章。如涉及带注释内容的数据（如 GenBank或EMBL格式文件）, 本章内容则非常重要。</p>
<p>尽管本章内容涵盖了 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 和 <tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 对象的大部分内容，但如需了解更多，读者可自行查阅 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> wiki (<a class="reference external" href="http://biopython.org/wiki/SeqRecord">http://biopython.org/wiki/SeqRecord</a> ),和内置帮助文档 (或在线文档 <a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SeqRecord.SeqRecord-class.html">SeqRecord</a> 和 <a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SeqFeature.SeqFeature-class.html">SeqFeature</a> )，获取更多信息:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">help</span><span class="p">(</span><span class="n">SeqRecord</span><span class="p">)</span>
<span class="gp">...</span>
</pre></div>
</div>
<div class="section" id="seqrecord">
<h2>4.1  SeqRecord对象<a class="headerlink" href="#seqrecord" title="Permalink to this headline">¶</a></h2>
<p><tt class="docutils literal"><span class="pre">SeqRecord</span></tt> (Sequence Record) 类包含在 <tt class="docutils literal"><span class="pre">Bio.SeqRecord</span></tt> 模块中。该类是 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 序列输入/输出交互界面 (详见Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>)的基本数据类型。可以把identifiers 和features等高级属性与序列关联起来 (参见Chapter <a class="reference external" href="#chapter:Bio.Seq">3</a>)。</p>
<p><tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 类非常简单,包括下列属性:</p>
<dl class="docutils">
<dt><strong>.seq</strong></dt>
<dd>– 序列自身（即 <tt class="docutils literal"><span class="pre">Seq</span></tt> 对象）。</dd>
<dt><strong>.id</strong></dt>
<dd>– 序列主ID（-字符串类型）。通常类同于accession number。</dd>
<dt><strong>.name</strong></dt>
<dd>– 序列名/id （-字符串类型）。 可以是accession number, 也可是clone名（类似GenBank record中的LOCUS id）。</dd>
<dt><strong>.description</strong></dt>
<dd>– 序列描述（-字符串类型）。</dd>
<dt><strong>.letter_annotations</strong></dt>
<dd>– 对照序列的每个字母逐字注释（per-letter-annotations），以信息名为键（keys），信息内容为值（value）所构成的字典。值与序列等长，用Python列表、元组或字符串表示。.letter_annotations可用于质量分数(如    Section <a class="reference external" href="#sec:FASTQ-filtering-example">18.1.6</a>) 或二级结构信息 (如 Stockholm/PFAM 比对文件)等数据的存储。</dd>
<dt><strong>.annotations</strong></dt>
<dd>– 用于储存附加信息的字典。信息名为键（keys），信息内容为值（value）。用于保存序列的零散信息（如unstructured information）。</dd>
<dt><strong>.features</strong></dt>
<dd>– <tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 对象列表，储存序列的结构化信息（structured information），如：基因位置, 蛋白结构域。features 详见本章第三节（ Section <a class="reference external" href="#sec:seq_features">4.3</a>）。</dd>
<dt><strong>.dbxrefs</strong></dt>
<dd>– 储存数据库交叉引用信息（cross-references）的字符串列表。</dd>
</dl>
</div>
<div class="section" id="id2">
<h2>4.2  创建 SeqRecord<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h2>
<p>使用 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 对象非常简单，因为所有的信息都存储在该类的属性中；通常不必手动新建，用 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 从序列文件读取即可（见Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>）。 当然新建 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 也不复杂。</p>
<div class="section" id="id3">
<h3>4.2.1  从头新建SeqRecord<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<p><tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 最少只需包含 <tt class="docutils literal"><span class="pre">Seq</span></tt> 对象:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">simple_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s">&quot;GATC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">simple_seq_r</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">simple_seq</span><span class="p">)</span>
</pre></div>
</div>
<p>还可以通过初始化函数给 id, name和description赋值；反之，它们被设为默认值“unknown”（可随后编辑）:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">simple_seq_r</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;&lt;unknown id&gt;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">simple_seq_r</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="s">&quot;AC12345&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">simple_seq_r</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="s">&quot;Made up sequence I wish I could write a paper about&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">simple_seq_r</span><span class="o">.</span><span class="n">description</span>
<span class="go">Made up sequence I wish I could write a paper about</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">simple_seq_r</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Seq(&#39;GATC&#39;, Alphabet())</span>
</pre></div>
</div>
<p>标识符对输出 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 内容到文件很重要，可随SeqRecord同时建立:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">simple_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s">&quot;GATC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">simple_seq_r</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">simple_seq</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="s">&quot;AC12345&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>上述章节已提到，<tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 含有一个 <tt class="docutils literal"><span class="pre">annotations</span></tt> 属性，用于储存各种杂乱注释的字典。添加annotations示例如下:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">simple_seq_r</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s">&quot;evidence&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;None. I just made it up.&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">simple_seq_r</span><span class="o">.</span><span class="n">annotations</span>
<span class="go">{&#39;evidence&#39;: &#39;None. I just made it up.&#39;}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">simple_seq_r</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s">&quot;evidence&quot;</span><span class="p">]</span>
<span class="go">None. I just made it up.</span>
</pre></div>
</div>
<p><tt class="docutils literal"><span class="pre">letter_annotations</span></tt> 也是字典，其值为与序列等长的内置Python字符串、列表或元组:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">simple_seq_r</span><span class="o">.</span><span class="n">letter_annotations</span><span class="p">[</span><span class="s">&quot;phred_quality&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">40</span><span class="p">,</span><span class="mi">40</span><span class="p">,</span><span class="mi">38</span><span class="p">,</span><span class="mi">30</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">simple_seq_r</span><span class="o">.</span><span class="n">letter_annotations</span>
<span class="go">{&#39;phred_quality&#39;: [40, 40, 38, 30]}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">simple_seq_r</span><span class="o">.</span><span class="n">letter_annotations</span><span class="p">[</span><span class="s">&quot;phred_quality&quot;</span><span class="p">]</span>
<span class="go">[40, 40, 38, 30]</span>
</pre></div>
</div>
<p><tt class="docutils literal"><span class="pre">dbxrefs</span></tt> 和 <tt class="docutils literal"><span class="pre">features</span></tt> 分别是字符串和 <tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 对象的Python列表，将在后续章节讨论。</p>
</div>
<div class="section" id="fastaseqrecord">
<h3>4.2.2  根据FASTA文件创建SeqRecord对象<a class="headerlink" href="#fastaseqrecord" title="Permalink to this headline">¶</a></h3>
<p>本节以鼠疫耶尔森菌株（<em>Yersinia pestis biovar Microtus</em> str. 91001 ）的pPCP1质粒全长序列为例,说明从FASTA文件创建SeqRecord的过程。该序列原始文件来自NCBI，可在Biopython单元测试GenBank文件夹下找到，也可点击 <a class="reference external" href="http://biopython.org/SRC/biopython/Tests/GenBank/NC_005816.fna">NC_005816.fna</a> 下载。</p>
<p>序列以大于号开头，该文件只包含一条序列:</p>
<div class="code python highlight-python"><pre>&gt;gi|45478711|ref|NC_005816.1| Yersinia pestis biovar Microtus ... pPCP1, complete sequence
TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGGGGGTAATCTGCTCTCC
...</pre>
</div>
<p>回顾 Chapter <a class="reference external" href="#chapter:quick-start">2</a> 的内容，我们已经遇到过 <tt class="docutils literal"><span class="pre">Bio.SeqIO.parse(...)</span></tt> 函数，用于遍历 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 对象中的所有记录。 此处，我们介绍 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 模块中的另一个类似函数Bio.SeqIO.read()，用于读取单条序列的文件 （详见 Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> ）:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&quot;NC_005816.fna&quot;</span><span class="p">,</span> <span class="s">&quot;fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span>
<span class="go">SeqRecord(seq=Seq(&#39;TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG&#39;,</span>
<span class="go">SingleLetterAlphabet()), id=&#39;gi|45478711|ref|NC_005816.1|&#39;, name=&#39;gi|45478711|ref|NC_005816.1|&#39;,</span>
<span class="go">description=&#39;gi|45478711|ref|NC_005816.1| Yersinia pestis biovar Microtus ... sequence&#39;,</span>
<span class="go">dbxrefs=[])</span>
</pre></div>
</div>
<p>现在让我们逐个介绍 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 对象中的主要属性，从给予我们序列属性的 <tt class="docutils literal"><span class="pre">Seq</span></tt> 对象 开始:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Seq(&#39;TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG&#39;, SingleLetterAlphabet())</span>
</pre></div>
</div>
<p>此处 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 默认为通用字母表（generic alphabet）, 而非判断是否DNA序列。如果FASTA文件中序列类型已知，也可通过 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 自行设定 (见Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> 用法)。</p>
<p>接下来介绍 identifiers 和 description:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;gi|45478711|ref|NC_005816.1|&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">name</span>
<span class="go">&#39;gi|45478711|ref|NC_005816.1|&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">description</span>
<span class="go">&#39;gi|45478711|ref|NC_005816.1| Yersinia pestis biovar Microtus ... pPCP1, complete sequence&#39;</span>
</pre></div>
</div>
<p>FASTA文件中序列名所在行的第一个单词(去除大于号后) 被当作 <tt class="docutils literal"><span class="pre">id</span></tt> 和 <tt class="docutils literal"><span class="pre">name</span></tt> ；而将整行 (去除大于号后) 作为 description。这样设定是为了向后兼容，同时也为了便于处理如下序列:</p>
<div class="code python highlight-python"><pre>&gt;Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1
TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGGGGGTAATCTGCTCTCC
...</pre>
</div>
<p>Note: 读取FASTA 文件时其他注释属性为空:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">dbxrefs</span>
<span class="go">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span>
<span class="go">{}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">letter_annotations</span>
<span class="go">{}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">features</span>
<span class="go">[]</span>
</pre></div>
</div>
<p>本例中FASTA文件源于NCBI，其规范的格式，意味着我们可以方便的解析这些信息并选择提取GI和accession number等信息。然后，对于从其他来源获得的FASTA文件，并不能确保能获得这些信息。</p>
</div>
<div class="section" id="genbank-seqrecord">
<h3>4.2.3  从 GenBank文件创建 SeqRecord<a class="headerlink" href="#genbank-seqrecord" title="Permalink to this headline">¶</a></h3>
<p>仍以疫耶尔森菌株pPCP1质粒全长序列（<em>Yersinia pestis biovar Microtus</em> str. 91001 plasmid pPCP1）为例，不同的是这次使用Genbank格式的文件，该文件同样包含在Biopython单元测试/GenBank文件夹下, 也可点击 <a class="reference external" href="http://biopython.org/SRC/biopython/Tests/GenBank/NC_005816.gb">NC_005816.gb</a>
下载。</p>
<p>该文件只含一条记录 (只有一个 LOCUS 行):</p>
<div class="code python highlight-python"><pre>LOCUS       NC_005816               9609 bp    DNA     circular BCT 21-JUL-2008
DEFINITION  Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete
            sequence.
ACCESSION   NC_005816
VERSION     NC_005816.1  GI:45478711
PROJECT     GenomeProject:10638
...</pre>
</div>
<p>同样使用 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 读取文件，代码跟处理FASTA 文件类似 (详见
Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> for details):</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&quot;NC_005816.gb&quot;</span><span class="p">,</span> <span class="s">&quot;genbank&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span>
<span class="go">SeqRecord(seq=Seq(&#39;TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG&#39;,</span>
<span class="go">IUPACAmbiguousDNA()), id=&#39;NC_005816.1&#39;, name=&#39;NC_005816&#39;,</span>
<span class="go">description=&#39;Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.&#39;,</span>
<span class="go">dbxrefs=[&#39;Project:10638&#39;])</span>
</pre></div>
</div>
<p>你可能已经发现了一些不同之处，逐个环顾各个属性，序列字符串和上述类似，但此处 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 可自动识别序列类型 （详见 chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> ）:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">seq</span>
<span class="go">Seq(&#39;TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG&#39;, IUPACAmbiguousDNA())</span>
</pre></div>
</div>
<p><tt class="docutils literal"><span class="pre">name</span></tt> 源于 LOCUS行, <tt class="docutils literal"><span class="pre">id</span></tt> 附加了版本后缀。description源于DEFINITION 行:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;NC_005816.1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">name</span>
<span class="go">&#39;NC_005816&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">description</span>
<span class="go">&#39;Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.&#39;</span>
</pre></div>
</div>
<p>GenBank 文件中per-letter annotations为空:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">letter_annotations</span>
<span class="go">{}</span>
</pre></div>
</div>
<p>多数注释信息储存在 <tt class="docutils literal"><span class="pre">annotations</span></tt> 字典中，例如:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">)</span>
<span class="go">11</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s">&quot;source&quot;</span><span class="p">]</span>
<span class="go">&#39;Yersinia pestis biovar Microtus str. 91001&#39;</span>
</pre></div>
</div>
<p><tt class="docutils literal"><span class="pre">dbxrefs</span></tt> 列表中的数据来自 PROJECT 或DBLINK行:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">dbxrefs</span>
<span class="go">[&#39;Project:10638&#39;]</span>
</pre></div>
</div>
<p>最后也许也可能是最有意思的， <tt class="docutils literal"><span class="pre">features</span></tt> 列表以 <tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 对象的形式保存了features table中的所有entries（如genes和CDS等）。</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
<span class="go">29</span>
</pre></div>
</div>
<p>接下来，我们将在 Section <a class="reference external" href="#sec:seq_features">4.3</a> 介绍 <tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 对象。</p>
</div>
</div>
<div class="section" id="feature-location-position">
<h2>4.3  Feature, location 和 position对象<a class="headerlink" href="#feature-location-position" title="Permalink to this headline">¶</a></h2>
<div class="section" id="seqfeature">
<h3>4.3.1  SeqFeature对象<a class="headerlink" href="#seqfeature" title="Permalink to this headline">¶</a></h3>
<p>序列特征是描述一条序列不可或缺的部分。抛开序列本身，你需要一种方式去组织和获取关于这条序列的 “抽象” 信息。 尽管设计一个通用的类囊括序列的所有特征看似是不可能的，但是Biopython的 <tt class="docutils literal"><span class="pre">SeqFeature</span></tt>
类试图尽可能多的囊括序列的所有特征。Biopython主要依据 GenBank/EMBL 特征表来设计相应的对象，认识到这一点，将有助于你更快更好的理解Biopython <tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 对象。</p>
<p><tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 对象的关键目的在于描述其相对于父序列（parent sequence，通常为 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 对象）所处的位置（location）, 通常是介于两个positions间的一个区域（region），后续Section <a class="reference external" href="#sec:locations">4.3.2</a> 将详细说明。</p>
<p><tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 对象含大量属性，首先一一例出，然后在后续章节举例说明其用法:</p>
<dl class="docutils">
<dt><strong>.type</strong></dt>
<dd>– 用文字描述的feature类型 (如 ‘CDS’ 或 ‘gene’)。</dd>
<dt><strong>.location</strong></dt>
<dd><p class="first">– <tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 在序列中所处的位置。见Section <a class="reference external" href="#sec:locations">4.3.2</a>。 <tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 设计了众多针对location对象的功能，包含一系列简写的属性。</p>
<dl class="last docutils">
<dt><strong>.ref</strong></dt>
<dd>– <tt class="docutils literal"><span class="pre">.location.ref</span></tt> 简写 &#8211;location对象相关的参考序列。通常为空（None）。</dd>
<dt><strong>.ref_db</strong></dt>
<dd>– <tt class="docutils literal"><span class="pre">.location.ref_db</span></tt> 简写 &#8211; 指定 <tt class="docutils literal"><span class="pre">.ref</span></tt> 相关数据库名称。通常为空（None）。</dd>
<dt><strong>.strand</strong></dt>
<dd>– <tt class="docutils literal"><span class="pre">.location.strand</span></tt> 简写 &#8211; 表示feature所处序列的strand。在双链核酸序列中，1表示正链, -1表示负链, 0 表示strand信息很重要但未知, None表示strand信息未知且不重要。蛋白和单链核酸序列为None。</dd>
</dl>
</dd>
<dt><strong>.qualifiers</strong></dt>
<dd>– 存储feature附加信息（Python字典）。键（key）为值（value）所存信息的单字简要描述，值为实际信息。比如，键为 “evidence” ，而值为 “computational (non-experimental)”。 这只是为了提醒人们注意，该feature没有被实验所证实（湿实验）。Note：为与GenBank/EMBL文件中的feature tables对应，规定.qualifiers 中值为字符串数组（即使只有一个字符串）。</dd>
<dt><strong>.sub_features</strong></dt>
<dd>– 只有在描述复杂位置时才使用，如 GenBank/EMBL文件中的 ‘joins’ 位置。 已被 <tt class="docutils literal"><span class="pre">CompoundLocation</span></tt> 对象取代，因此略过不提。</dd>
</dl>
</div>
<div class="section" id="positionslocations">
<h3>4.3.2  Positions和locations<a class="headerlink" href="#positionslocations" title="Permalink to this headline">¶</a></h3>
<p><tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 对象主要用于描述相对于父序列中的位置（region）信息。Region用location对象表示，通常是两个position间的范围。为了区分location和position，我们定义如下:</p>
<dl class="docutils">
<dt><strong>position</strong></dt>
<dd>– 表示位于序列中的单一位置, 可以是精确的也可以是不确定的位置（如5, 20, <tt class="docutils literal"><span class="pre">&lt;100</span></tt> 和 <tt class="docutils literal"><span class="pre">&gt;200</span></tt> ）。</dd>
<dt><strong>location</strong></dt>
<dd>– 介于两个positions间的区域。比如5..20 (5到20)。</dd>
</dl>
<p>之所以特意提及这两个概念是因为我经常混淆两者。</p>
<div class="section" id="featurelocation">
<h4>4.3.2.1  FeatureLocation 对象<a class="headerlink" href="#featurelocation" title="Permalink to this headline">¶</a></h4>
<p>多数 <tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 特别简单（真核基因例外），只需起点、终点以及strand信息。最基本的 <tt class="docutils literal"><span class="pre">FeatureLocation</span></tt> 对象中通常包括上述三点信息。</p>
<p>但实际情况未必如此简单，因为我们还需处理包含几个区域的复合locations，而且position本身很可能是不精确的。</p>
</div>
<div class="section" id="compoundlocation">
<h4>4.3.2.2  CompoundLocation 对象<a class="headerlink" href="#compoundlocation" title="Permalink to this headline">¶</a></h4>
<p>为了更方便的处理EMBL/GenBank文件中的 ‘join’ locations，Biopython 1.62引入 <tt class="docutils literal"><span class="pre">CompoundLocation</span></tt> 对象。</p>
</div>
<div class="section" id="positions">
<h4>4.3.2.3  模糊Positions<a class="headerlink" href="#positions" title="Permalink to this headline">¶</a></h4>
<p>目前，我们只处理过简单position，feature location复杂因素之一就是由position本身
不准确所致。生物学中许多问题都是不确定的，比如：你通过双核苷酸priming证明了mRNA
的转录起始位点是这两个位点中的一个。这是十分有价值的发现，但困难来自于怎样表述这
个位点信息。为了处理类似情况，我们用模糊位点（fuzzy position）表示。根据fuzzy
position的不同，我们用5个类分别描述:</p>
<dl class="docutils">
<dt><strong>ExactPosition</strong></dt>
<dd>– 精确位点，用一个数字表示。从该对象的 <tt class="docutils literal"><span class="pre">position</span></tt> 属性可得知精确位点信息。</dd>
<dt><strong>BeforePosition</strong></dt>
<dd>– 位于某个特定位点前。如 <tt class="docutils literal"><span class="pre">`&lt;13'</span></tt> , 在GenBank/EMBL中代表实际位点位于13之前。从该对象的 <tt class="docutils literal"><span class="pre">position</span></tt> 属性可得知上边界信息。</dd>
<dt><strong>AfterPosition</strong></dt>
<dd>– 与 <tt class="docutils literal"><span class="pre">BeforePosition</span></tt> 相反,如 <tt class="docutils literal"><span class="pre">`&gt;13'</span></tt> , 在GenBank/EMBL中代表实际位点位于13以后。从该对象的 <tt class="docutils literal"><span class="pre">position</span></tt> 属性可获知下边界信息。</dd>
<dt><strong>WithinPosition</strong></dt>
<dd>– 介于两个特定位点之间，偶尔在GenBank/EMBL locations用到。如 ‘(1.5)’, GenBank/EMBL中代表实际位点位于1到5之间。该对象需要两个position属性表示，第一个 <tt class="docutils literal"><span class="pre">position</span></tt> 表示下边界（本例为1）， <tt class="docutils literal"><span class="pre">extension</span></tt> 表示上边界与下边界的差值（本例为4）。因此在WithinPosition中， <tt class="docutils literal"><span class="pre">object.position</span></tt> 表示下边界， <tt class="docutils literal"><span class="pre">object.position</span> <span class="pre">+</span> <span class="pre">object.extension</span></tt> 表示上边界。</dd>
<dt><strong>OneOfPosition</strong></dt>
<dd>– 表示几个位点中的一个（GenBank/EMBL文件中偶尔能看到），比如在基因起始位点不明确或者有两个候选位点的时候可以使用，或者用于明确表示两个相关基因特征时使用。</dd>
<dt><strong>UnknownPosition</strong></dt>
<dd>– 代表未知位点。在GenBank/EMBL文件中没有使用，对应 UniProt中的 ‘?’ feature坐标。</dd>
</dl>
<p>举例说明创建一个fuzzy end points:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqFeature</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">start_pos</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="o">.</span><span class="n">AfterPosition</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">end_pos</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="o">.</span><span class="n">BetweenPosition</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="n">left</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">right</span><span class="o">=</span><span class="mi">9</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_location</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="n">start_pos</span><span class="p">,</span> <span class="n">end_pos</span><span class="p">)</span>
</pre></div>
</div>
<p>Note：Biopython 1.59以后，fuzzy-locations有修改, 特别是BetweenPosition和
WithinPosition，现在必须显示用整数表示。起点为较小值，终点则为较大值。</p>
<p>print输出 <tt class="docutils literal"><span class="pre">FeatureLocation</span></tt> 对象，可看到简洁的结果:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">my_location</span>
<span class="go">[&gt;5:(8^9)]</span>
</pre></div>
</div>
<p>也可通过start和end属性得到fuzzy position的起始/终止位点:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">my_location</span><span class="o">.</span><span class="n">start</span>
<span class="go">AfterPosition(5)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">my_location</span><span class="o">.</span><span class="n">start</span>
<span class="go">&gt;5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_location</span><span class="o">.</span><span class="n">end</span>
<span class="go">BetweenPosition(9, left=8, right=9)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">my_location</span><span class="o">.</span><span class="n">end</span>
<span class="go">(8^9)</span>
</pre></div>
</div>
<p>如果你只想获取数字，不理会模糊positions，则可将fuzzy position强制转换成一个整数:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">int</span><span class="p">(</span><span class="n">my_location</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">int</span><span class="p">(</span><span class="n">my_location</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
<span class="go">9</span>
</pre></div>
</div>
<p>为了兼容旧版Biopython，保留了整数形式的 <tt class="docutils literal"><span class="pre">nofuzzy_start</span></tt> and <tt class="docutils literal"><span class="pre">nofuzzy_end</span></tt> :</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">my_location</span><span class="o">.</span><span class="n">nofuzzy_start</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_location</span><span class="o">.</span><span class="n">nofuzzy_end</span>
<span class="go">9</span>
</pre></div>
</div>
<p>Notice：上述例子只是为了帮助你理解fuzzy locations。</p>
<p>相似的，如果要建立一个精确location，只需将整数传递给 <tt class="docutils literal"><span class="pre">FeaturePosition</span></tt>
构造函数, 即可建立 <tt class="docutils literal"><span class="pre">ExactPosition</span></tt> 对象:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">exact_location</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="o">.</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">9</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">exact_location</span>
<span class="go">[5:9]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exact_location</span><span class="o">.</span><span class="n">start</span>
<span class="go">ExactPosition(5)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">int</span><span class="p">(</span><span class="n">exact_location</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">exact_location</span><span class="o">.</span><span class="n">nofuzzy_start</span>
<span class="go">5</span>
</pre></div>
</div>
<p>以上是Biopython处理fuzzy position的实现方法。希望读者能体会之所以这样设计，都是为了使用上的方便（至少不比精确位点复杂）</p>
</div>
<div class="section" id="location-testing">
<h4>4.3.2.4  Location testing<a class="headerlink" href="#location-testing" title="Permalink to this headline">¶</a></h4>
<p>可用Python关键词 <tt class="docutils literal"><span class="pre">in</span></tt> 检验某个碱基或氨基酸残基的父坐标是否位于
feature/location中。</p>
<p>假定你想知道某个SNP位于哪个feature里，并知道该SNP的索引位置是4350（Python 计数）。一个简单的实现方案是用循环遍历所有features:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_snp</span> <span class="o">=</span> <span class="mi">4350</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&quot;NC_005816.gb&quot;</span><span class="p">,</span> <span class="s">&quot;genbank&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">my_snp</span> <span class="ow">in</span> <span class="n">feature</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">print</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span><span class="p">,</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&#39;db_xref&#39;</span><span class="p">)</span>
<span class="gp">...</span>
<span class="go">source [&#39;taxon:229193&#39;]</span>
<span class="go">gene [&#39;GeneID:2767712&#39;]</span>
<span class="go">CDS [&#39;GI:45478716&#39;, &#39;GeneID:2767712&#39;]</span>
</pre></div>
</div>
<p>Note： GenBank /EMBL 文件中的 gene 和CDS features（ <tt class="docutils literal"><span class="pre">join</span></tt> ）只包含外显子，不含内含子。</p>
</div>
</div>
<div class="section" id="feature-location">
<h3>4.3.3  使用feature 或 location描述序列<a class="headerlink" href="#feature-location" title="Permalink to this headline">¶</a></h3>
<p><tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 或 location object对象并没有直接包含任何序列，只是可根据储存的location (见Section <a class="reference external" href="#sec:locations">4.3.2</a>)，从父序列中取得。例如：某一短基因位于负链5:18位置，由于GenBank/EMBL文件以1开始计数，Biopython中表示为 <tt class="docutils literal"><span class="pre">complement(6..18)</span></tt> :</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqFeature</span> <span class="kn">import</span> <span class="n">SeqFeature</span><span class="p">,</span> <span class="n">FeatureLocation</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">example_parent</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s">&quot;ACCGAGACGGCAAAGGCTAGCATAGGTATGAGACTTCCTTCCTGCCAGTGCTGAGGAACTGGGAGCCTAC&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">example_feature</span> <span class="o">=</span> <span class="n">SeqFeature</span><span class="p">(</span><span class="n">FeatureLocation</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">18</span><span class="p">),</span> <span class="nb">type</span><span class="o">=</span><span class="s">&quot;gene&quot;</span><span class="p">,</span> <span class="n">strand</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
<p>你可以用切片从父序列截取5:18,然后取反向互补序列。如果是Biopython 1.59或以后版本，可使用如下方法:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">feature_seq</span> <span class="o">=</span> <span class="n">example_parent</span><span class="p">[</span><span class="n">example_feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">start</span><span class="p">:</span><span class="n">example_feature</span><span class="o">.</span><span class="n">location</span><span class="o">.</span><span class="n">end</span><span class="p">]</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">feature_seq</span>
<span class="go">AGCCTTTGCCGTC</span>
</pre></div>
</div>
<p>不过在处理复合 features (joins)时，此法相当繁琐。此时可以使用 <tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 对象的 <tt class="docutils literal"><span class="pre">extract</span></tt> 方法处理:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">feature_seq</span> <span class="o">=</span> <span class="n">example_feature</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">example_parent</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">feature_seq</span>
<span class="go">AGCCTTTGCCGTC</span>
</pre></div>
</div>
<p><tt class="docutils literal"><span class="pre">SeqFeature</span></tt> 或 location对象的长度等同于所表示序列的长度。</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">example_feature</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">example_parent</span><span class="p">)</span>
<span class="go">AGCCTTTGCCGTC</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">example_feature</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">example_parent</span><span class="p">))</span>
<span class="go">13</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">example_feature</span><span class="p">)</span>
<span class="go">13</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="nb">len</span><span class="p">(</span><span class="n">example_feature</span><span class="o">.</span><span class="n">location</span><span class="p">)</span>
<span class="go">13</span>
</pre></div>
</div>
<p>简单 <tt class="docutils literal"><span class="pre">FeatureLocation</span></tt> 对象的长度等于终止osition减去起始position的差值；而 <tt class="docutils literal"><span class="pre">CompoundLocation</span></tt> 的长度则为各片段长度之和。</p>
</div>
</div>
<div class="section" id="references">
<h2>4.4  References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<p>对一条序列的注释还包括参考文献（reference），Biopython通过
<tt class="docutils literal"><span class="pre">Bio.SeqFeature.Reference</span></tt> 对象来储存相关的文献信息。</p>
<p>References属性储存了 <tt class="docutils literal"><span class="pre">期刊名</span></tt> 、 <tt class="docutils literal"><span class="pre">题名</span></tt> 、 <tt class="docutils literal"><span class="pre">作者</span></tt> 等信息。此外还包括 <tt class="docutils literal"><span class="pre">medline_id</span></tt> 、 <tt class="docutils literal"><span class="pre">pubmed_id</span></tt> 以及 <tt class="docutils literal"><span class="pre">comment</span></tt> 。</p>
<p>通常reference 也有 <tt class="docutils literal"><span class="pre">location</span></tt> 对象，便于文献涉及研究对象在序列中的定位。该 <tt class="docutils literal"><span class="pre">location</span></tt> 有可能是一个fuzzy location（见 section <a class="reference external" href="#sec:locations">4.3.2</a>）。</p>
<p>文献对象都以列表储存在 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 对象的 <tt class="docutils literal"><span class="pre">annotations</span></tt> 字典中。 字典的键为 “references”。reference对象也是为了方便处理文献而设计，希望能满足各种使用需求。</p>
</div>
<div class="section" id="id4">
<h2>4.5  格式化方法<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h2>
<p><tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 类中的 <tt class="docutils literal"><span class="pre">format()</span></tt> 能将字符串转换成被 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 支持的格式，如FASTA:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">generic_protein</span>

<span class="n">record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s">&quot;MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD&quot;</span> \
                      <span class="o">+</span><span class="s">&quot;GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK&quot;</span> \
                      <span class="o">+</span><span class="s">&quot;NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM&quot;</span> \
                      <span class="o">+</span><span class="s">&quot;SSAC&quot;</span><span class="p">,</span> <span class="n">generic_protein</span><span class="p">),</span>
                   <span class="nb">id</span><span class="o">=</span><span class="s">&quot;gi|14150838|gb|AAK54648.1|AF376133_1&quot;</span><span class="p">,</span>
                   <span class="n">description</span><span class="o">=</span><span class="s">&quot;chalcone synthase [Cucumis sativus]&quot;</span><span class="p">)</span>

<span class="k">print</span> <span class="n">record</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s">&quot;fasta&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>输出为:</p>
<div class="code python highlight-python"><pre>&gt;gi|14150838|gb|AAK54648.1|AF376133_1 chalcone synthase [Cucumis sativus]
MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD
GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK
NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM
SSAC</pre>
</div>
<p><tt class="docutils literal"><span class="pre">format</span></tt> 方法接收单个必选参数，小写字母字符串是 <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> 模块支持的输出格式 (见Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>)。然而，此 <tt class="docutils literal"><span class="pre">format()</span></tt> 方法并不适用于包含多条序列的文件格式 (如多序列比对格式)（详见Section <a class="reference external" href="#sec:Bio.SeqIO-and-StringIO">5.5.4</a>）。</p>
</div>
<div class="section" id="id5">
<h2>4.6  SeqRecord切片<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h2>
<p>通过切片截取 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 的部分序列可得到一条新的 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 。此处需引起注意的是per-letter annotations也被取切片, 但新序列中的features保持不变 (locations相应调整)。</p>
<p>以前述Genbank文件为例:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&quot;NC_005816.gb&quot;</span><span class="p">,</span> <span class="s">&quot;genbank&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span>
<span class="go">SeqRecord(seq=Seq(&#39;TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG&#39;,</span>
<span class="go">IUPACAmbiguousDNA()), id=&#39;NC_005816.1&#39;, name=&#39;NC_005816&#39;,</span>
<span class="go">description=&#39;Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.&#39;,</span>
<span class="go">dbxrefs=[&#39;Project:10638&#39;])</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="go">9609</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
<span class="go">41</span>
</pre></div>
</div>
<p>本例中，我们关注 <tt class="docutils literal"><span class="pre">YP_pPCP05</span></tt> 质粒上的 <tt class="docutils literal"><span class="pre">pim</span></tt> 基因。从GenBank文件可直接看出 <tt class="docutils literal"><span class="pre">pim</span></tt> gene/CDS location是 <tt class="docutils literal"><span class="pre">4343..4780</span></tt> （相应的Python 位置是 <tt class="docutils literal"><span class="pre">4342:4780</span></tt> ）。Location信息位于GenBank文件第12和13 entries中, 由于python以0开始计数，因此python中，它们是 <tt class="docutils literal"><span class="pre">features</span></tt> 列表中的 entries 11和12:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="mi">20</span><span class="p">]</span>
<span class="go">type: gene</span>
<span class="go">location: [4342:4780](+)</span>
<span class="go">qualifiers:</span>
<span class="go">    Key: db_xref, Value: [&#39;GeneID:2767712&#39;]</span>
<span class="go">    Key: gene, Value: [&#39;pim&#39;]</span>
<span class="go">    Key: locus_tag, Value: [&#39;YP_pPCP05&#39;]</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="mi">21</span><span class="p">]</span>
<span class="go">type: CDS</span>
<span class="go">location: [4342:4780](+)</span>
<span class="go">qualifiers:</span>
<span class="go">    Key: codon_start, Value: [&#39;1&#39;]</span>
<span class="go">    Key: db_xref, Value: [&#39;GI:45478716&#39;, &#39;GeneID:2767712&#39;]</span>
<span class="go">    Key: gene, Value: [&#39;pim&#39;]</span>
<span class="go">    Key: locus_tag, Value: [&#39;YP_pPCP05&#39;]</span>
<span class="go">    Key: note, Value: [&#39;similar to many previously sequenced pesticin immunity ...&#39;]</span>
<span class="go">    Key: product, Value: [&#39;pesticin immunity protein&#39;]</span>
<span class="go">    Key: protein_id, Value: [&#39;NP_995571.1&#39;]</span>
<span class="go">    Key: transl_table, Value: [&#39;11&#39;]</span>
<span class="go">    Key: translation, Value: [&#39;MGGGMISKLFCLALIFLSSSGLAEKNTYTAKDILQNLELNTFGNSLSH...&#39;]</span>
</pre></div>
</div>
<p>从父记录中取切片（4300 到 4800），观测所得到的features数量:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sub_record</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="mi">4300</span><span class="p">:</span><span class="mi">4800</span><span class="p">]</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sub_record</span>
<span class="go">SeqRecord(seq=Seq(&#39;ATAAATAGATTATTCCAAATAATTTATTTATGTAAGAACAGGATGGGAGGGGGA...TTA&#39;,</span>
<span class="go">IUPACAmbiguousDNA()), id=&#39;NC_005816.1&#39;, name=&#39;NC_005816&#39;,</span>
<span class="go">description=&#39;Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.&#39;,</span>
<span class="go">dbxrefs=[])</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">sub_record</span><span class="p">)</span>
<span class="go">500</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">sub_record</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
<span class="go">2</span>
</pre></div>
</div>
<p>子记录（sub_record）只包括两个features, 分别是 <tt class="docutils literal"><span class="pre">YP_pPCP05</span></tt> 质粒的gene和CDS:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">sub_record</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">type: gene</span>
<span class="go">location: [42:480](+)</span>
<span class="go">qualifiers:</span>
<span class="go">    Key: db_xref, Value: [&#39;GeneID:2767712&#39;]</span>
<span class="go">    Key: gene, Value: [&#39;pim&#39;]</span>
<span class="go">    Key: locus_tag, Value: [&#39;YP_pPCP05&#39;]</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">sub_record</span><span class="o">.</span><span class="n">features</span><span class="p">[</span><span class="mi">20</span><span class="p">]</span>
<span class="go">type: CDS</span>
<span class="go">location: [42:480](+)</span>
<span class="go">qualifiers:</span>
<span class="go">    Key: codon_start, Value: [&#39;1&#39;]</span>
<span class="go">    Key: db_xref, Value: [&#39;GI:45478716&#39;, &#39;GeneID:2767712&#39;]</span>
<span class="go">    Key: gene, Value: [&#39;pim&#39;]</span>
<span class="go">    Key: locus_tag, Value: [&#39;YP_pPCP05&#39;]</span>
<span class="go">    Key: note, Value: [&#39;similar to many previously sequenced pesticin immunity ...&#39;]</span>
<span class="go">    Key: product, Value: [&#39;pesticin immunity protein&#39;]</span>
<span class="go">    Key: protein_id, Value: [&#39;NP_995571.1&#39;]</span>
<span class="go">    Key: transl_table, Value: [&#39;11&#39;]</span>
<span class="go">    Key: translation, Value: [&#39;MGGGMISKLFCLALIFLSSSGLAEKNTYTAKDILQNLELNTFGNSLSH...&#39;]</span>
</pre></div>
</div>
<p>注意：locations已被调整至对应生成的新父序列!</p>
<p>尽可能灵敏和直观地获取子记录的相关特征（和任意的per-letter annotation），但是对于其余注释，Biopython无法判断是否仍然适用于子记录。因此子记录忽略了 <tt class="docutils literal"><span class="pre">annotations</span></tt> 和 <tt class="docutils literal"><span class="pre">dbxrefs</span></tt> 以避免引起歧义。</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sub_record</span><span class="o">.</span><span class="n">annotations</span>
<span class="go">{}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sub_record</span><span class="o">.</span><span class="n">dbxrefs</span>
<span class="go">[]</span>
</pre></div>
</div>
<p>为了便于实际操作，子记录保留了 <tt class="docutils literal"><span class="pre">id</span></tt> , <tt class="docutils literal"><span class="pre">name</span></tt> 和 <tt class="docutils literal"><span class="pre">description</span></tt> :</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sub_record</span><span class="o">.</span><span class="n">id</span>
<span class="go">&#39;NC_005816.1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sub_record</span><span class="o">.</span><span class="n">name</span>
<span class="go">&#39;NC_005816&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sub_record</span><span class="o">.</span><span class="n">description</span>
<span class="go">&#39;Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.&#39;</span>
</pre></div>
</div>
<p>上述例子很好的展示了问题，由于子记录不包括完整的质粒序列，因此description是错的。我们可以将子记录看做是截短版的GenBank文件，可用Section <cite>4.5 &lt;#sec:SeqRecord-format&gt;`__中所述 ``format`</cite> 方法纠正：
:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">sub_record</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="s">&quot;Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, partial.&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">sub_record</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s">&quot;genbank&quot;</span><span class="p">)</span>
<span class="gp">...</span>
</pre></div>
</div>
<p>FASTQ例子参见 Sections <a class="reference external" href="#sec:FASTQ-slicing-off-primer">18.1.7</a>
和 <a class="reference external" href="#sec:FASTQ-slicing-off-adaptor">18.1.8</a> （此例中per-letter annotations (read质量分数) 也被取切片）。</p>
</div>
<div class="section" id="id6">
<h2>4.7  SeqRecord对象相加<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h2>
<p><tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 对象可相加得到一个新的 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 。注意：per-letter annotations也相加, features (locations 调整)；而其它annotation 保持不变(如id、name和description)。</p>
<p>以FASTQ 文件中的第一条记录为例说明per-letter annotation （Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> 详细介绍 <tt class="docutils literal"><span class="pre">SeqIO</span></tt> 函数）:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s">&quot;example.fastq&quot;</span><span class="p">,</span> <span class="s">&quot;fastq&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">next</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="go">25</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">record</span><span class="o">.</span><span class="n">seq</span>
<span class="go">CCCTTCTTGTCTTCAGCGTTTCTCC</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">record</span><span class="o">.</span><span class="n">letter_annotations</span><span class="p">[</span><span class="s">&quot;phred_quality&quot;</span><span class="p">]</span>
<span class="go">[26, 26, 18, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 22, 26, 26, 26, 26,</span>
<span class="go">26, 26, 26, 23, 23]</span>
</pre></div>
</div>
<p>假设上述序列数据来自Roche 454测序, 你根据其它信息得知 <tt class="docutils literal"><span class="pre">TTT</span></tt> 应该是 <tt class="docutils literal"><span class="pre">TT</span></tt> 。此时可分别用切片提取第三个 <tt class="docutils literal"><span class="pre">T</span></tt> 前后的序列（ <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> ）:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">left</span> <span class="o">=</span> <span class="n">record</span><span class="p">[:</span><span class="mi">20</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">left</span><span class="o">.</span><span class="n">seq</span>
<span class="go">CCCTTCTTGTCTTCAGCGTT</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">left</span><span class="o">.</span><span class="n">letter_annotations</span><span class="p">[</span><span class="s">&quot;phred_quality&quot;</span><span class="p">]</span>
<span class="go">[26, 26, 18, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 22, 26, 26, 26, 26]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">right</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="mi">21</span><span class="p">:]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">right</span><span class="o">.</span><span class="n">seq</span>
<span class="go">CTCC</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">right</span><span class="o">.</span><span class="n">letter_annotations</span><span class="p">[</span><span class="s">&quot;phred_quality&quot;</span><span class="p">]</span>
<span class="go">[26, 26, 23, 23]</span>
</pre></div>
</div>
<p>两部分相加:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">edited</span> <span class="o">=</span> <span class="n">left</span> <span class="o">+</span> <span class="n">right</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">edited</span><span class="p">)</span>
<span class="go">24</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">edited</span><span class="o">.</span><span class="n">seq</span>
<span class="go">CCCTTCTTGTCTTCAGCGTTCTCC</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">edited</span><span class="o">.</span><span class="n">letter_annotations</span><span class="p">[</span><span class="s">&quot;phred_quality&quot;</span><span class="p">]</span>
<span class="go">[26, 26, 18, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 26, 22, 26, 26, 26, 26,</span>
<span class="go">26, 26, 23, 23]</span>
</pre></div>
</div>
<p>很容易和直观吧！上述两步可合并:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">edited</span> <span class="o">=</span> <span class="n">record</span><span class="p">[:</span><span class="mi">20</span><span class="p">]</span> <span class="o">+</span> <span class="n">record</span><span class="p">[</span><span class="mi">21</span><span class="p">:]</span>
</pre></div>
</div>
<p>现在以GenBank文件（假定是环状基因组）为例说明features:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&quot;NC_005816.gb&quot;</span><span class="p">,</span> <span class="s">&quot;genbank&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span>
<span class="go">SeqRecord(seq=Seq(&#39;TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG&#39;,</span>
<span class="go">IUPACAmbiguousDNA()), id=&#39;NC_005816.1&#39;, name=&#39;NC_005816&#39;,</span>
<span class="go">description=&#39;Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.&#39;,</span>
<span class="go">dbxrefs=[&#39;Project:10638&#39;])</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="go">9609</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
<span class="go">41</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">dbxrefs</span>
<span class="go">[&#39;Project:58037&#39;]</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">[&#39;comment&#39;, &#39;sequence_version&#39;, &#39;source&#39;, &#39;taxonomy&#39;, &#39;keywords&#39;, &#39;references&#39;,</span>
<span class="go">&#39;accessions&#39;, &#39;data_file_division&#39;, &#39;date&#39;, &#39;organism&#39;, &#39;gi&#39;]</span>
</pre></div>
</div>
<p>可改变起点:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">shifted</span> <span class="o">=</span> <span class="n">record</span><span class="p">[</span><span class="mi">2000</span><span class="p">:]</span> <span class="o">+</span> <span class="n">record</span><span class="p">[:</span><span class="mi">2000</span><span class="p">]</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">shifted</span>
<span class="go">SeqRecord(seq=Seq(&#39;GATACGCAGTCATATTTTTTACACAATTCTCTAATCCCGACAAGGTCGTAGGTC...GGA&#39;,</span>
<span class="go">IUPACAmbiguousDNA()), id=&#39;NC_005816.1&#39;, name=&#39;NC_005816&#39;,</span>
<span class="go">description=&#39;Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.&#39;,</span>
<span class="go">dbxrefs=[])</span>
</pre></div>
</div>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">shifted</span><span class="p">)</span>
<span class="go">9609</span>
</pre></div>
</div>
<p>Note: 上述方法并不完美（丢失了数据库交叉引用dbxrefs 和源feature）:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">shifted</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
<span class="go">40</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">shifted</span><span class="o">.</span><span class="n">dbxrefs</span>
<span class="go">[]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">shifted</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">[]</span>
</pre></div>
</div>
<p>这是因为 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 切片对 annotation 保留非常谨慎 (错误保留 annotation 可能引起大问题)。如果你想保留数据库的交叉引用dbxrefs和其余annotations 字典必须明确说明，才能得以保留:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">shifted</span><span class="o">.</span><span class="n">dbxrefs</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">dbxrefs</span><span class="p">[:]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">shifted</span><span class="o">.</span><span class="n">annotations</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">shifted</span><span class="o">.</span><span class="n">dbxrefs</span>
<span class="go">[&#39;Project:10638&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">shifted</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="go">[&#39;comment&#39;, &#39;sequence_version&#39;, &#39;source&#39;, &#39;taxonomy&#39;, &#39;keywords&#39;, &#39;references&#39;,</span>
<span class="go">&#39;accessions&#39;, &#39;data_file_division&#39;, &#39;date&#39;, &#39;organism&#39;, &#39;gi&#39;]</span>
</pre></div>
</div>
<p>Note: 此例中序列record的identifiers也应调整（因为NCBI的reference链接的是未经修改的 <em>原始</em> 序列）。</p>
</div>
<div class="section" id="id7">
<h2>4.8  反向互补SeqRecord对象<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h2>
<p>为消除序列反向互补后annotation改变带来的困难，Biopython 1.57 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 对象加入了 <tt class="docutils literal"><span class="pre">reverse_complement</span></tt> 方法。这也成为Biopython 1.57的新特性之一。</p>
<p>序列用Seq对象中的reverse_complement方法反向互补。Features随location而改变，strand也被重新计算。复制并反转per-letter-annotation（通常情况下这种做法比较合适，如对质量分数注释的反转）。然而多数annotation的转变却存有问题。</p>
<p>比如record ID是accession号，该accession不应被用于反向互补序列。默认identifier转换可导致后续分析中的轻度数据损坏。因此 <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 的id、
name、description、annotations和dbxrefs默认不变。</p>
<p><tt class="docutils literal"><span class="pre">SeqRecord</span></tt> 对象的 <tt class="docutils literal"><span class="pre">reverse_complement</span></tt> 法用多个可选参数以对应record的属性。将这些参数设为 <tt class="docutils literal"><span class="pre">True</span></tt> 表示复制旧值；而 <tt class="docutils literal"><span class="pre">False</span></tt> 意为用缺省值替换旧值。当然也可自定义新值。</p>
<p>举例:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s">&quot;NC_005816.gb&quot;</span><span class="p">,</span> <span class="s">&quot;genbank&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">dbxrefs</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">)</span>
<span class="go">NC_005816.1 9609 41 1 11</span>
</pre></div>
</div>
<p>反向互补该record并给ID赋予新值 - 注意：多数annotation丢失，而features仍在:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">rc</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">(</span><span class="nb">id</span><span class="o">=</span><span class="s">&quot;TESTING&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">rc</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">rc</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">rc</span><span class="o">.</span><span class="n">features</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">rc</span><span class="o">.</span><span class="n">dbxrefs</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">rc</span><span class="o">.</span><span class="n">annotations</span><span class="p">)</span>
<span class="go">TESTING 9609 41 0 0</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">第4章  序列注释对象</a><ul>
<li><a class="reference internal" href="#seqrecord">4.1  SeqRecord对象</a></li>
<li><a class="reference internal" href="#id2">4.2  创建 SeqRecord</a><ul>
<li><a class="reference internal" href="#id3">4.2.1  从头新建SeqRecord</a></li>
<li><a class="reference internal" href="#fastaseqrecord">4.2.2  根据FASTA文件创建SeqRecord对象</a></li>
<li><a class="reference internal" href="#genbank-seqrecord">4.2.3  从 GenBank文件创建 SeqRecord</a></li>
</ul>
</li>
<li><a class="reference internal" href="#feature-location-position">4.3  Feature, location 和 position对象</a><ul>
<li><a class="reference internal" href="#seqfeature">4.3.1  SeqFeature对象</a></li>
<li><a class="reference internal" href="#positionslocations">4.3.2  Positions和locations</a><ul>
<li><a class="reference internal" href="#featurelocation">4.3.2.1  FeatureLocation 对象</a></li>
<li><a class="reference internal" href="#compoundlocation">4.3.2.2  CompoundLocation 对象</a></li>
<li><a class="reference internal" href="#positions">4.3.2.3  模糊Positions</a></li>
<li><a class="reference internal" href="#location-testing">4.3.2.4  Location testing</a></li>
</ul>
</li>
<li><a class="reference internal" href="#feature-location">4.3.3  使用feature 或 location描述序列</a></li>
</ul>
</li>
<li><a class="reference internal" href="#references">4.4  References</a></li>
<li><a class="reference internal" href="#id4">4.5  格式化方法</a></li>
<li><a class="reference internal" href="#id5">4.6  SeqRecord切片</a></li>
<li><a class="reference internal" href="#id6">4.7  SeqRecord对象相加</a></li>
<li><a class="reference internal" href="#id7">4.8  反向互补SeqRecord对象</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="chr03.html"
                        title="previous chapter">第3章     生物序列对象</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="chr05.html"
                        title="next chapter">第5章  序列输入和输出</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/chr04.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="chr05.html" title="第5章 序列输入和输出"
             >next</a> |</li>
        <li class="right" >
          <a href="chr03.html" title="第3章 生物序列对象"
             >previous</a> |</li>
        <li><a href="index.html">biopython_cn 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, biopythoners.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.
    </div>
  </body>
</html>