<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>References &mdash; biopython_cn 0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="biopython_cn 0.1 documentation" href="index.html" />
    <link rel="prev" title="第22章 附录：Python相关知识" href="chr22.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="chr22.html" title="第22章 附录：Python相关知识"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">biopython_cn 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="references">
<h1>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h1>
<dl class="docutils" id="cock2009">
<dt>[1]</dt>
<dd>Peter J. A. Cock, Tiago Antao, Jeffrey T. Chang, Brad A. Chapman,
Cymon J. Cox, Andrew Dalke, Iddo Friedberg, Thomas Hamelryck, Frank
Kauff, Bartek Wilczynski, Michiel J. L. de Hoon: “Biopython: freely
available Python tools for computational molecular biology and
bioinformatics”. <em>Bioinformatics</em> <strong>25</strong> (11), 1422–1423 (2009).
<a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/btp163">doi:10.1093/bioinformatics/btp163</a>,</dd>
</dl>
<dl class="docutils" id="pritchard2006">
<dt>[2]</dt>
<dd>Leighton Pritchard, Jennifer A. White, Paul R.J. Birch, Ian K. Toth:
“GenomeDiagram: a python package for the visualization of
large-scale genomic data”. <em>Bioinformatics</em> <strong>22</strong> (5): 616–617
(2006).
<a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/btk021">doi:10.1093/bioinformatics/btk021</a>,</dd>
</dl>
<dl class="docutils" id="toth2006">
<dt>[3]</dt>
<dd>Ian K. Toth, Leighton Pritchard, Paul R. J. Birch: “Comparative
genomics reveals what makes an enterobacterial plant pathogen”.
<em>Annual Review of Phytopathology</em> <strong>44</strong>: 305–336 (2006).
<a class="reference external" href="http://dx.doi.org/10.1146/annurev.phyto.44.070505.143444">doi:10.1146/annurev.phyto.44.070505.143444</a>,</dd>
</dl>
<dl class="docutils" id="vanderauwera2009">
<dt>[4]</dt>
<dd>Géraldine A. van der Auwera, Jaroslaw E. Król, Haruo Suzuki, Brian
Foster, Rob van Houdt, Celeste J. Brown, Max Mergeay, Eva M. Top:
“Plasmids captured in C. metallidurans CH34: defining the PromA
family of broad-host-range plasmids”. <em>Antonie van Leeuwenhoek</em>
<strong>96</strong> (2): 193–204 (2009).
<a class="reference external" href="http://dx.doi.org/10.1007/s10482-009-9316-9">doi:10.1007/s10482-009-9316-9</a></dd>
</dl>
<dl class="docutils" id="proux2002">
<dt>[5]</dt>
<dd>Caroline Proux, Douwe van Sinderen, Juan Suarez, Pilar Garcia,
Victor Ladero, Gerald F. Fitzgerald, Frank Desiere, Harald Brüssow:
“The dilemma of phage taxonomy illustrated by comparative genomics
of Sfi21-Like Siphoviridae in lactic acid bacteria”. <em>Journal of
Bacteriology</em> <strong>184</strong> (21): 6026–6036 (2002).
<a class="reference external" href="http://dx.doi.org/10.1128/JB.184.21.6026-6036.2002">http://dx.doi.org/10.1128/JB.184.21.6026-6036.2002</a></dd>
</dl>
<dl class="docutils" id="jupe2012">
<dt>[6]</dt>
<dd>Florian Jupe, Leighton Pritchard, Graham J. Etherington, Katrin
MacKenzie, Peter JA Cock, Frank Wright, Sanjeev Kumar Sharma1, Dan
Bolser, Glenn J Bryan, Jonathan DG Jones, Ingo Hein: “Identification
and localisation of the NB-LRR gene family within the potato
genome”. <em>BMC Genomics</em> <strong>13</strong>: 75 (2012).
<a class="reference external" href="http://dx.doi.org/10.1186/1471-2164-13-75">http://dx.doi.org/10.1186/1471-2164-13-75</a></dd>
</dl>
<dl class="docutils" id="cock2010">
<dt>[7]</dt>
<dd>Peter J. A. Cock, Christopher J. Fields, Naohisa Goto, Michael L.
Heuer, Peter M. Rice: “The Sanger FASTQ file format for sequences
with quality scores, and the Solexa/Illumina FASTQ variants”.
<em>Nucleic Acids Research</em> <strong>38</strong> (6): 1767–1771 (2010).
<a class="reference external" href="http://dx.doi.org/10.1093/nar/gkp1137">doi:10.1093/nar/gkp1137</a></dd>
<dt>[8]</dt>
<dd>Patrick O. Brown, David Botstein: “Exploring the new world of the
genome with DNA microarrays”. <em>Nature Genetics</em> <strong>21</strong> (Supplement
1), 33–37 (1999).
<a class="reference external" href="http://dx.doi.org/10.1038/4462">doi:10.1038/4462</a></dd>
</dl>
<dl class="docutils" id="talevich2012">
<dt>[9]</dt>
<dd>Eric Talevich, Brandon M. Invergo, Peter J.A. Cock, Brad A. Chapman:
“Bio.Phylo: A unified toolkit for processing, analyzing and
visualizing phylogenetic trees in Biopython”. <em>BMC Bioinformatics</em>
<strong>13</strong>: 209 (2012).
<a class="reference external" href="http://dx.doi.org/10.1186/1471-2105-13-209">doi:10.1186/1471-2105-13-209</a></dd>
</dl>
<dl class="docutils" id="cornish1985">
<dt>[10]</dt>
<dd>Athel Cornish-Bowden: “Nomenclature for incompletely specified bases
in nucleic acid sequences: Recommendations 1984.” <em>Nucleic Acids
Research</em> <strong>13</strong> (9): 3021–3030 (1985).
<a class="reference external" href="http://dx.doi.org/10.1093/nar/13.9.3021">doi:10.1093/nar/13.9.3021</a></dd>
</dl>
<dl class="docutils" id="cavener1987">
<dt>[11]</dt>
<dd>Douglas R. Cavener: “Comparison of the consensus sequence flanking
translational start sites in Drosophila and vertebrates.” <em>Nucleic
Acids Research</em> <strong>15</strong> (4): 1353–1361 (1987).
<a class="reference external" href="http://dx.doi.org/10.1093/nar/15.4.1353">doi:10.1093/nar/15.4.1353</a></dd>
</dl>
<dl class="docutils" id="bailey1994">
<dt>[12]</dt>
<dd>Timothy L. Bailey and Charles Elkan: “Fitting a mixture model by
expectation maximization to discover motifs in biopolymers”,
<em>Proceedings of the Second International Conference on Intelligent
Systems for Molecular Biology</em> 28–36. AAAI Press, Menlo Park,
California (1994).</dd>
</dl>
<dl class="docutils" id="chapman2000">
<dt>[13]</dt>
<dd>Brad Chapman and Jeff Chang: “Biopython: Python tools for
computational biology”. <em>ACM SIGBIO Newsletter</em> <strong>20</strong> (2): 15–19
(August 2000).</dd>
</dl>
<dl class="docutils" id="dehoon2004">
<dt>[14]</dt>
<dd>Michiel J. L. de Hoon, Seiya Imoto, John Nolan, Satoru Miyano: “Open
source clustering software”. <em>Bioinformatics</em> <strong>20</strong> (9): 1453–1454
(2004).
<a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/bth078">doi:10.1093/bioinformatics/bth078</a></dd>
<dt>[15]</dt>
<dd>Michiel B. Eisen, Paul T. Spellman, Patrick O. Brown, David
Botstein: “Cluster analysis and display of genome-wide expression
patterns”. <em>Proceedings of the National Academy of Science USA</em>
<strong>95</strong> (25): 14863–14868 (1998).
<a class="reference external" href="http://dx.doi.org/10.1073/pnas.96.19.10943-c">doi:10.1073/pnas.96.19.10943-c</a></dd>
</dl>
<dl class="docutils" id="golub1971">
<dt>[16]</dt>
<dd>Gene H. Golub, Christian Reinsch: “Singular value decomposition and
least squares solutions”. In <em>Handbook for Automatic Computation</em>,
<strong>2</strong>, (Linear Algebra) (J. H. Wilkinson and C. Reinsch, eds),
134–151. New York: Springer-Verlag (1971).</dd>
</dl>
<dl class="docutils" id="golub1989">
<dt>[17]</dt>
<dd>Gene H. Golub, Charles F. Van Loan: <em>Matrix computations</em>, 2nd
edition (1989).</dd>
</dl>
<dl class="docutils" id="hamelryck2003a">
<dt>[18]</dt>
<dd>Thomas Hamelryck and Bernard Manderick: 11PDB parser and structure
class implemented in Python”. <em>Bioinformatics</em>, <strong>19</strong> (17):
2308–2310 (2003) <a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/btg299">doi:
10.1093/bioinformatics/btg299</a>.</dd>
</dl>
<dl class="docutils" id="hamelryck2003b">
<dt>[19]</dt>
<dd>Thomas Hamelryck: “Efficient identification of side-chain patterns
using a multidimensional index tree”. <em>Proteins</em> <strong>51</strong> (1): 96–108
(2003).
<a class="reference external" href="http://dx.doi.org/10.1002/prot.10338">doi:10.1002/prot.10338</a></dd>
</dl>
<dl class="docutils" id="hamelryck2005">
<dt>[20]</dt>
<dd>Thomas Hamelryck: “An amino acid has two sides; A new 2D measure
provides a different view of solvent exposure”. <em>Proteins</em> <strong>59</strong>
(1): 29–48 (2005).
<a class="reference external" href="http://dx.doi.org/10.1002/prot.20379">doi:10.1002/prot.20379</a>.</dd>
<dt>[21]</dt>
<dd>John A. Hartiga. <em>Clustering algorithms</em>. New York: Wiley (1975).</dd>
<dt>[22]</dt>
<dd>Anil L. Jain, Richard C. Dubes: <em>Algorithms for clustering data</em>.
Englewood Cliffs, N.J.: Prentice Hall (1988).</dd>
</dl>
<dl class="docutils" id="kachitvichyanukul1988">
<dt>[23]</dt>
<dd>Voratas Kachitvichyanukul, Bruce W. Schmeiser: Binomial Random
Variate Generation. <em>Communications of the ACM</em> <strong>31</strong> (2): 216–222
(1988).
<a class="reference external" href="http://dx.doi.org/10.1145/42372.42381">doi:10.1145/42372.42381</a></dd>
</dl>
<dl class="docutils" id="kohonen1997">
<dt>[24]</dt>
<dd>Teuvo Kohonen: “Self-organizing maps”, 2nd Edition. Berlin; New
York: Springer-Verlag (1997).</dd>
</dl>
<dl class="docutils" id="lecuyer1988">
<dt>[25]</dt>
<dd>Pierre L’Ecuyer: “Efficient and Portable Combined Random Number
Generators.” <em>Communications of the ACM</em> <strong>31</strong> (6): 742–749,774
(1988).
<a class="reference external" href="http://dx.doi.org/10.1145/62959.62969">doi:10.1145/62959.62969</a></dd>
</dl>
<dl class="docutils" id="majumdar2005">
<dt>[26]</dt>
<dd>Indraneel Majumdar, S. Sri Krishna, Nick V. Grishin: “PALSSE: A
program to delineate linear secondary structural elements from
protein structures.” <em>BMC Bioinformatics</em>, <strong>6</strong>: 202 (2005).
<a class="reference external" href="http://dx.doi.org/10.1186/1471-2105-6-202">doi:10.1186/1471-2105-6-202</a>.</dd>
</dl>
<dl class="docutils" id="matys2003">
<dt>[27]</dt>
<dd>V. Matys, E. Fricke, R. Geffers, E. G?ssling, M. Haubrock, R. Hehl,
K. Hornischer, D. Karas, A.E. Kel, O.V. Kel-Margoulis, D.U. Kloos,
S. Land, B. Lewicki-Potapov, H. Michael, R. Münch, I. Reuter, S.
Rotert, H. Saxel, M. Scheer, S. Thiele, E. Wingender E: “TRANSFAC:
transcriptional regulation, from patterns to profiles.” Nucleic
Acids Research <strong>31</strong> (1): 374–378 (2003).
<a class="reference external" href="http://dx.doi.org/10.1093/nar/gkg108">doi:10.1093/nar/gkg108</a></dd>
<dt>[28]</dt>
<dd>Robin Sibson: “SLINK: An optimally efficient algorithm for the
single-link cluster method”. <em>The Computer Journal</em> <strong>16</strong> (1):
30–34 (1973).
<a class="reference external" href="http://dx.doi.org/10.1093/comjnl/16.1.30">doi:10.1093/comjnl/16.1.30</a></dd>
</dl>
<dl class="docutils" id="snedecor1989">
<dt>[29]</dt>
<dd>George W. Snedecor, William G. Cochran: <em>Statistical methods</em>. Ames,
Iowa: Iowa State University Press (1989).</dd>
</dl>
<dl class="docutils" id="tamayo1999">
<dt>[30]</dt>
<dd>Pablo Tamayo, Donna Slonim, Jill Mesirov, Qing Zhu, Sutisak
Kitareewan, Ethan Dmitrovsky, Eric S. Lander, Todd R. Golub:
“Interpreting patterns of gene expression with self-organizing maps:
Methods and application to hematopoietic differentiation”.
<em>Proceedings of the National Academy of Science USA</em> <strong>96</strong> (6):
2907–2912 (1999).
<a class="reference external" href="http://dx.doi.org/10.1073/pnas.96.6.2907">doi:10.1073/pnas.96.6.2907</a></dd>
<dt>[31]</dt>
<dd>Robert C. Tryon, Daniel E. Bailey: <em>Cluster analysis</em>. New York:
McGraw-Hill (1970).</dd>
<dt>[32]</dt>
<dd>John W. Tukey: “Exploratory data analysis”. Reading, Mass.:
Addison-Wesley Pub. Co. (1977).</dd>
</dl>
<dl class="docutils" id="yeung2001">
<dt>[33]</dt>
<dd>Ka Yee Yeung, Walter L. Ruzzo: “Principal Component Analysis for
clustering gene expression data”. <em>Bioinformatics</em> <strong>17</strong> (9):
763–774 (2001).
<a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/17.9.763">doi:10.1093/bioinformatics/17.9.763</a></dd>
<dt>[34]</dt>
<dd>Alok Saldanha: “Java Treeview—extensible visualization of microarray
data”. <em>Bioinformatics</em> <strong>20</strong> (17): 3246–3248 (2004).
<a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/bth349">http://dx.doi.org/10.1093/bioinformatics/bth349</a></dd>
</dl>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="chr22.html"
                        title="previous chapter">第22章 附录：Python相关知识</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/chr23.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="chr22.html" title="第22章 附录：Python相关知识"
             >previous</a> |</li>
        <li><a href="index.html">biopython_cn 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, biopythoners.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.
    </div>
  </body>
</html>